{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "248b240e-8393-4934-bb7f-4e6fb192bb1a",
   "metadata": {},
   "source": [
    "Get cumulative numbers of variants, targets and genes for hicov geographic cohorts \n",
    "and hicov all "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51459662-9dab-4e66-90bb-56048c62d873",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef9cfdd1-2070-466a-9ef8-e42aee245a1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import allel\n",
    "import collections\n",
    "import numpy as np\n",
    "import malariagen_data\n",
    "import gcsfs\n",
    "import zarr\n",
    "import dask\n",
    "import dask.array as da\n",
    "import cython\n",
    "import functools\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6b0137e-0db0-4898-b4c1-2168a95a9779",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from dask.distributed import Client\n",
    "import dask\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': False}) # Silence large chunk warnings\n",
    "import dask.array as da\n",
    "from dask import delayed, compute\n",
    "from dask_gateway import Gateway\n",
    "import functools\n",
    "import numcodecs\n",
    "from fsspec.implementations.zip import ZipFileSystem\n",
    "from collections.abc import Mapping\n",
    "import gcsfs\n",
    "import numba\n",
    "import psutil\n",
    "from humanize import naturalsize\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f425e08-f531-401a-ae5b-53550cdece7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gcs = gcsfs.GCSFileSystem()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cbfd441-21c4-496b-9795-7c6a50547fb5",
   "metadata": {},
   "source": [
    "## Read ref genome, accessibility and annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b92ce2e-901c-47f0-9aad-35f12fa9dcdd",
   "metadata": {},
   "source": [
    "### Ref genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7da25f9c-fea3-415a-8631-3610622dcf32",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.1.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Load af1 to access reference genome\n",
    "#Sequence will be read for chromosomes separately\n",
    "af1 = malariagen_data.Af1(release='1.0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4e5f02-8503-4751-8914-95408a3c29c4",
   "metadata": {},
   "source": [
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "858a6eb6-cf0a-43c7-a7e8-816384803252",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(468233,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">ncRNA_gene</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000001</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">tRNA</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000001.R2</td><td style=\"text-align: center\">AFUN2_000001</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">exon</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">exon_Trnak-cuu-2_t1-E1</td><td style=\"text-align: center\">AFUN2_000001.R2</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">468230</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">46442600</td><td style=\"text-align: center\">46443004</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AFUN2_014819.P27467-CDS2</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">468231</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">five_prime_UTR</td><td style=\"text-align: center\">46441974</td><td style=\"text-align: center\">46442325</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AFUN2_014819.R27467_1</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">468232</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">46443005</td><td style=\"text-align: center\">46443055</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AFUN2_014819.R27467_2</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(468233,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AfunGA1_2RL', 'VEuPathDB', 'ncRNA_gene', 44146272, 44146344, -1., '-', -1, 'AFUN2_000001', '.')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'tRNA', 44146272, 44146344, -1., '-', -1, 'AFUN2_000001.R2', 'AFUN2_000001')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'exon', 44146272, 44146344, -1., '-', -1, 'exon_Trnak-cuu-2_t1-E1', 'AFUN2_000001.R2')\n",
       " ...\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'CDS', 46442600, 46443004, -1., '+',  0, 'AFUN2_014819.P27467-CDS2', 'AFUN2_014819.R27467')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'five_prime_UTR', 46441974, 46442325, -1., '+', -1, 'utr_AFUN2_014819.R27467_1', 'AFUN2_014819.R27467')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'three_prime_UTR', 46443005, 46443055, -1., '+', -1, 'utr_AFUN2_014819.R27467_2', 'AFUN2_014819.R27467')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Uploaded local copy of annotation\n",
    "#Because the release contains a preliminary version\n",
    "gff_fn = 'VectorBase-65_AfunestusAfunGA1.gff'\n",
    "features = allel.FeatureTable.from_gff3(gff_fn, attributes=['ID', 'Parent'])\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a1aca07e-2693-4dbb-9f39-fa5401b6825b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('exon', 193386),\n",
       " ('CDS', 161982),\n",
       " ('five_prime_UTR', 42469),\n",
       " ('three_prime_UTR', 26305),\n",
       " ('mRNA', 25342),\n",
       " ('protein_coding_gene', 12092),\n",
       " ('ncRNA_gene', 2280),\n",
       " ('lnc_RNA', 1474),\n",
       " ('rRNA', 1011),\n",
       " ('ncRNA', 646),\n",
       " ('pseudogene', 447),\n",
       " ('pseudogenic_transcript', 447),\n",
       " ('tRNA', 316),\n",
       " ('snRNA', 27),\n",
       " ('snoRNA', 9)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Eyeball the features present\n",
    "collections.Counter(features.type).most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c2c430b5-462c-4992-b8c2-47db9851acca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AFUN2_000001'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# index features by their parent\n",
    "idx_feature_parent = dict(features[['ID', 'Parent']])\n",
    "idx_feature_parent['AFUN2_000001.R2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4474ab23-2a97-4ec4-83b8-e1d0648b4c31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('AfunGA1_2RL', 224647),\n",
       " ('AfunGA1_3RL', 188399),\n",
       " ('AfunGA1_X', 45230),\n",
       " ('CALSEJ010000020', 656),\n",
       " ('CALSEJ010000036', 305),\n",
       " ('CALSEJ010000073', 279),\n",
       " ('CALSEJ010000056', 223),\n",
       " ('CALSEJ010000119', 207),\n",
       " ('CALSEJ010000194', 204),\n",
       " ('CALSEJ010000306', 203)]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check scaffold names\n",
    "collections.Counter(features.seqid).most_common()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6da04d08-0214-4803-9ae2-6da30b89c992",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(458276,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">ncRNA_gene</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000001</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">tRNA</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000001.R2</td><td style=\"text-align: center\">AFUN2_000001</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">exon</td><td style=\"text-align: center\">44146272</td><td style=\"text-align: center\">44146344</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">exon_Trnak-cuu-2_t1-E1</td><td style=\"text-align: center\">AFUN2_000001.R2</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">458273</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">46442600</td><td style=\"text-align: center\">46443004</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AFUN2_014819.P27467-CDS2</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">458274</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">five_prime_UTR</td><td style=\"text-align: center\">46441974</td><td style=\"text-align: center\">46442325</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AFUN2_014819.R27467_1</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">458275</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">46443005</td><td style=\"text-align: center\">46443055</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AFUN2_014819.R27467_2</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(458276,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AfunGA1_2RL', 'VEuPathDB', 'ncRNA_gene', 44146272, 44146344, -1., '-', -1, 'AFUN2_000001', '.')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'tRNA', 44146272, 44146344, -1., '-', -1, 'AFUN2_000001.R2', 'AFUN2_000001')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'exon', 44146272, 44146344, -1., '-', -1, 'exon_Trnak-cuu-2_t1-E1', 'AFUN2_000001.R2')\n",
       " ...\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'CDS', 46442600, 46443004, -1., '+',  0, 'AFUN2_014819.P27467-CDS2', 'AFUN2_014819.R27467')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'five_prime_UTR', 46441974, 46442325, -1., '+', -1, 'utr_AFUN2_014819.R27467_1', 'AFUN2_014819.R27467')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'three_prime_UTR', 46443005, 46443055, -1., '+', -1, 'utr_AFUN2_014819.R27467_2', 'AFUN2_014819.R27467')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to features annotated on the chromosomes\n",
    "loc_features_chroms = (\n",
    "    (features.seqid == 'AfunGA1_2RL') |\n",
    "    (features.seqid == 'AfunGA1_3RL') |\n",
    "    (features.seqid == 'AfunGA1_X') \n",
    ")\n",
    "features_chroms = features[loc_features_chroms]\n",
    "features_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ff8f7c4-3859-4496-8f9e-0ebe7f98a5d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AfunGA1_2RL', 'AfunGA1_3RL', 'AfunGA1_X'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check that it worked\n",
    "np.unique(features_chroms.seqid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd61ae32-87bb-42f2-85fd-d1083cbaa3b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(161215,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AfunGA1_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">16057118</td><td style=\"text-align: center\">16058505</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AFUN2_000317.P27478-CDS9</td><td style=\"text-align: center\">AFUN2_000317.R27478</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AfunGA1_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">16057118</td><td style=\"text-align: center\">16058505</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AFUN2_000317.P27477-CDS9</td><td style=\"text-align: center\">AFUN2_000317.R27477</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AfunGA1_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">16057118</td><td style=\"text-align: center\">16058505</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AFUN2_000317.P27471-CDS8</td><td style=\"text-align: center\">AFUN2_000317.R27471</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">161212</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">22228717</td><td style=\"text-align: center\">22228942</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">1</td><td style=\"text-align: center\">AFUN2_014818.P27463-CDS9</td><td style=\"text-align: center\">AFUN2_014818.R27463</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">161213</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">46442326</td><td style=\"text-align: center\">46442514</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AFUN2_014819.P27467-CDS1</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">161214</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">46442600</td><td style=\"text-align: center\">46443004</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AFUN2_014819.P27467-CDS2</td><td style=\"text-align: center\">AFUN2_014819.R27467</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(161215,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AfunGA1_X', 'VEuPathDB', 'CDS', 16057118, 16058505, -1., '-', 2, 'AFUN2_000317.P27478-CDS9', 'AFUN2_000317.R27478')\n",
       " ('AfunGA1_X', 'VEuPathDB', 'CDS', 16057118, 16058505, -1., '-', 2, 'AFUN2_000317.P27477-CDS9', 'AFUN2_000317.R27477')\n",
       " ('AfunGA1_X', 'VEuPathDB', 'CDS', 16057118, 16058505, -1., '-', 2, 'AFUN2_000317.P27471-CDS8', 'AFUN2_000317.R27471')\n",
       " ...\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'CDS', 22228717, 22228942, -1., '+', 1, 'AFUN2_014818.P27463-CDS9', 'AFUN2_014818.R27463')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'CDS', 46442326, 46442514, -1., '+', 0, 'AFUN2_014819.P27467-CDS1', 'AFUN2_014819.R27467')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'CDS', 46442600, 46443004, -1., '+', 0, 'AFUN2_014819.P27467-CDS2', 'AFUN2_014819.R27467')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset to coding sequences on the chromosomes\n",
    "cdss_chroms = features_chroms[features_chroms.type == 'CDS']\n",
    "cdss_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d476e08e-2ea2-4987-82df-2d4d7bc6a5b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(11962,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AfunGA1_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">16056398</td><td style=\"text-align: center\">16128681</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000317</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">28414528</td><td style=\"text-align: center\">28429433</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000318</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">4506748</td><td style=\"text-align: center\">4514391</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_000319</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">11959</th><td style=\"text-align: center\">AfunGA1_2RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">33077479</td><td style=\"text-align: center\">33078614</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_014817</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">11960</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">22217243</td><td style=\"text-align: center\">22232561</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_014818</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">11961</th><td style=\"text-align: center\">AfunGA1_3RL</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">46441974</td><td style=\"text-align: center\">46443055</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AFUN2_014819</td><td style=\"text-align: center\">.</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(11962,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AfunGA1_X', 'VEuPathDB', 'protein_coding_gene', 16056398, 16128681, -1., '-', -1, 'AFUN2_000317', '.')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'protein_coding_gene', 28414528, 28429433, -1., '+', -1, 'AFUN2_000318', '.')\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'protein_coding_gene',  4506748,  4514391, -1., '-', -1, 'AFUN2_000319', '.')\n",
       " ...\n",
       " ('AfunGA1_2RL', 'VEuPathDB', 'protein_coding_gene', 33077479, 33078614, -1., '-', -1, 'AFUN2_014817', '.')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'protein_coding_gene', 22217243, 22232561, -1., '+', -1, 'AFUN2_014818', '.')\n",
       " ('AfunGA1_3RL', 'VEuPathDB', 'protein_coding_gene', 46441974, 46443055, -1., '+', -1, 'AFUN2_014819', '.')]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset to protein coding genes on the chromosomes\n",
    "genes_coding_chroms = features_chroms[features_chroms.type =='protein_coding_gene']\n",
    "genes_coding_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "10bfc7c9-597f-4f22-a454-ecf9b935b6b9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11962"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double-check that you get the same number of protein coding genes via cds\n",
    "uq_genes_coding_chroms = np.unique([idx_feature_parent[t] for t in cdss_chroms.Parent])\n",
    "assert set(genes_coding_chroms.ID) == set(uq_genes_coding_chroms)\n",
    "len(uq_genes_coding_chroms)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1480579-92b4-4d15-a648-67707ab45748",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2fc8acc3-54e2-427f-93d1-c5d3fe4253ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext cython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b49d53f3-54fd-4281-8e59-0b8ae4b36934",
   "metadata": {},
   "source": [
    "Don't know how to safely use comments in cython - so I'll include them here\n",
    "\n",
    "`%%cython` instructs to compile this cell with cython. the `-a` flag tells it to showcase the compilation.  \n",
    "Next define the nucleotides as integers corresponding to their ascii value.  \n",
    "`@cython.boundscheck(False)` tells cython to not perform a boundscheck -- so we have to guarantee the bounds.  \n",
    "`opt_is_cas9_target()` takes as arguments a 'view' of seq (which we explicitly say cannot be None), a 'view' of is_variant and a boolean specifying whether to check reverse complements. I guess a view is an dynamic array?  \n",
    "Initialise the `out` array, a counter `i` and boleans `fwd`, `rev` and fill `out` with as many zeroes as `seq` is long (but now make it a 1bit integer, whereas it was initialised as an 8bit integer?)  \n",
    "Then release the GIL (it's something of a global lock on python code, and it has to be released to allow for multithreading).  \n",
    "Loop through the positions in seq, taking into account the length of the targetsite. Check that of 21 consecutive positions, none are `N` (except the first of the PAM, which can be anything) and it ends in `GG`. Returns `fwd = True` if this check is satisfied. In that case, that that `is_variant` is zero at all positions except the third last (so if we give a boolean array of where bases are segregating within the population, it gets targets that are non-segregating; but if we give a boolean array of where bases are non-accessbile, it gets targets that are accessible). Repeat in reverse complement if `revcomp = True`.  \n",
    "After finishing the loop, return a view of an array of booleans specifiying for each target starting postition whether its a valid target either forwards or reverse. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4268b4ec-87cf-4c5f-b340-8482d6b69c27",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "\n",
    "import numpy as np\n",
    "cimport numpy as cnp\n",
    "cimport cython\n",
    "\n",
    "cdef:\n",
    "    cnp.uint8_t A = ord('A')\n",
    "    cnp.uint8_t C = ord('C')\n",
    "    cnp.uint8_t G = ord('G')\n",
    "    cnp.uint8_t T = ord('T')\n",
    "    cnp.uint8_t N = ord('N')\n",
    "    \n",
    "@cython.boundscheck(False)\n",
    "def opt_is_cas9_target(cnp.uint8_t[:] seq not None, cnp.uint8_t[:] is_variant, bint revcomp=True):\n",
    "    \"\"\"Locate Cas9 target sequences.\"\"\"\n",
    "    \n",
    "    cdef:\n",
    "        cnp.uint8_t[:] out\n",
    "        Py_ssize_t i\n",
    "        bint fwd, rev\n",
    "        \n",
    "    out = np.zeros(seq.shape[0], dtype='u1')\n",
    "    \n",
    "    with nogil:\n",
    "        for i in range(seq.shape[0] - 20):\n",
    "            # check if forward strand has -NGG\n",
    "            fwd = ((seq[i] != N) and\n",
    "                   (seq[i+1] != N) and\n",
    "                   (seq[i+2] != N) and\n",
    "                   (seq[i+3] != N) and\n",
    "                   (seq[i+4] != N) and\n",
    "                   (seq[i+5] != N) and\n",
    "                   (seq[i+6] != N) and\n",
    "                   (seq[i+7] != N) and\n",
    "                   (seq[i+8] != N) and\n",
    "                   (seq[i+9] != N) and\n",
    "                   (seq[i+10] != N) and\n",
    "                   (seq[i+11] != N) and\n",
    "                   (seq[i+12] != N) and\n",
    "                   (seq[i+13] != N) and\n",
    "                   (seq[i+14] != N) and\n",
    "                   (seq[i+15] != N) and\n",
    "                   (seq[i+16] != N) and\n",
    "                   (seq[i+17] != N) and\n",
    "                   # anything and\n",
    "                   (seq[i+19] == G) and \n",
    "                   (seq[i+20] == G))\n",
    "            if fwd and is_variant is not None:\n",
    "                fwd = (fwd and \n",
    "                       (is_variant[i] == 0) and\n",
    "                       (is_variant[i+1] == 0) and\n",
    "                       (is_variant[i+2] == 0) and\n",
    "                       (is_variant[i+3] == 0) and\n",
    "                       (is_variant[i+4] == 0) and\n",
    "                       (is_variant[i+5] == 0) and\n",
    "                       (is_variant[i+6] == 0) and\n",
    "                       (is_variant[i+7] == 0) and\n",
    "                       (is_variant[i+8] == 0) and\n",
    "                       (is_variant[i+9] == 0) and\n",
    "                       (is_variant[i+10] == 0) and\n",
    "                       (is_variant[i+11] == 0) and\n",
    "                       (is_variant[i+12] == 0) and\n",
    "                       (is_variant[i+13] == 0) and\n",
    "                       (is_variant[i+14] == 0) and\n",
    "                       (is_variant[i+15] == 0) and\n",
    "                       (is_variant[i+16] == 0) and\n",
    "                       (is_variant[i+17] == 0) and\n",
    "                       # anything and\n",
    "                       (is_variant[i+19] == 0) and\n",
    "                       (is_variant[i+20] == 0))\n",
    "            if revcomp:\n",
    "                # check if reverse strand has -NGG\n",
    "                rev = ((seq[i] == C) and\n",
    "                       (seq[i+1] == C) and\n",
    "                       # anything and\n",
    "                       (seq[i+3] != N) and\n",
    "                       (seq[i+4] != N) and\n",
    "                       (seq[i+5] != N) and\n",
    "                       (seq[i+6] != N) and\n",
    "                       (seq[i+7] != N) and\n",
    "                       (seq[i+8] != N) and\n",
    "                       (seq[i+9] != N) and\n",
    "                       (seq[i+10] != N) and\n",
    "                       (seq[i+11] != N) and\n",
    "                       (seq[i+12] != N) and\n",
    "                       (seq[i+13] != N) and\n",
    "                       (seq[i+14] != N) and\n",
    "                       (seq[i+15] != N) and\n",
    "                       (seq[i+16] != N) and\n",
    "                       (seq[i+17] != N) and\n",
    "                       (seq[i+18] != N) and\n",
    "                       (seq[i+19] != N) and \n",
    "                       (seq[i+20] != N))\n",
    "                if rev and is_variant is not None:\n",
    "                    rev = (rev and \n",
    "                           (is_variant[i] == 0) and\n",
    "                           (is_variant[i+1] == 0) and\n",
    "                           # anything and\n",
    "                           (is_variant[i+3] == 0) and\n",
    "                           (is_variant[i+4] == 0) and\n",
    "                           (is_variant[i+5] == 0) and\n",
    "                           (is_variant[i+6] == 0) and\n",
    "                           (is_variant[i+7] == 0) and\n",
    "                           (is_variant[i+8] == 0) and\n",
    "                           (is_variant[i+9] == 0) and\n",
    "                           (is_variant[i+10] == 0) and\n",
    "                           (is_variant[i+11] == 0) and\n",
    "                           (is_variant[i+12] == 0) and\n",
    "                           (is_variant[i+13] == 0) and\n",
    "                           (is_variant[i+14] == 0) and\n",
    "                           (is_variant[i+15] == 0) and\n",
    "                           (is_variant[i+16] == 0) and\n",
    "                           (is_variant[i+17] == 0) and\n",
    "                           (is_variant[i+18] == 0) and\n",
    "                           (is_variant[i+19] == 0) and\n",
    "                           (is_variant[i+20] == 0))\n",
    "            else:\n",
    "                rev = False\n",
    "            out[i] = fwd or rev\n",
    "    \n",
    "    return np.asarray(out).view(bool)\n",
    "\n",
    "@cython.boundscheck(False)\n",
    "def opt_all_subsequent(cython.integral[:] t, Py_ssize_t n):\n",
    "    \"\"\"Locate contiguous regions with the same non-zero value.\"\"\"\n",
    "    \n",
    "    cdef:\n",
    "        cython.integral[:] out\n",
    "        Py_ssize_t i, j\n",
    "        cython.integral x\n",
    "        bint subs\n",
    "        \n",
    "    out = np.zeros_like(t)\n",
    "    \n",
    "    with nogil:\n",
    "        for i in range(t.shape[0] - n):\n",
    "            x = t[i]\n",
    "            if x > 0:\n",
    "                subs = True\n",
    "                for j in range(i+1, i+n):\n",
    "                    subs = subs and t[j] == x\n",
    "                if subs:\n",
    "                    out[i] = x\n",
    "                \n",
    "    return np.asarray(out)\n",
    "  \n",
    "@cython.boundscheck(False)\n",
    "def opt_zero_subsequent(cython.integral[:] t, Py_ssize_t n):\n",
    "    \"\"\"Locate non-overlapping targets.\"\"\"\n",
    "\n",
    "    cdef:\n",
    "        cython.integral[:] out\n",
    "        Py_ssize_t i, j\n",
    "        cython.integral x\n",
    "        \n",
    "    out = np.zeros_like(t)\n",
    "    i = 0\n",
    "\n",
    "    with nogil:\n",
    "        while i < t.shape[0]:\n",
    "            x = t[i]\n",
    "            if x > 0:\n",
    "                out[i] = x\n",
    "                i += n\n",
    "            else:\n",
    "                i += 1\n",
    "\n",
    "    return np.asarray(out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d10e753-c0f1-45c8-8814-51ba3a5b965e",
   "metadata": {},
   "source": [
    "`opt_all_subsequent()` scans an array of integer values and reports for each entry, whether the `n` subsequent entries have the same non-zero value (used to check e.g. whether all entries are in the same exon). If the requirement is met, it replaces the corresponding entry in `out` by its value in `t`. Returns numpy array `out`, with zeroes where the condition is not met.  \n",
    "`opt_zero_subsequent()` scans an array of integer values and returns an array of integers where non-zero values are spaces apart by at least `n`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "866cea0d-cb76-4f71-a114-be949460f777",
   "metadata": {},
   "source": [
    "### Python functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9f66d875-86ba-45e2-abb0-244f0714227d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_reference_sequence(chrom):\n",
    "    \"\"\"Load the reference sequence.\"\"\"\n",
    "    seq = af1.genome_sequence(chrom).compute()\n",
    "    seq = np.char.upper(seq)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bb9b2a46-79f3-4fd1-ad86-07ef1858eaec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# can't cache, but it's pretty quick\n",
    "def get_is_cas9_target(chrom, revcomp=True, is_variant=None):\n",
    "    \"\"\"Find Cas9 target sequences.\"\"\"\n",
    "    seq = get_reference_sequence(chrom).view('u1')\n",
    "    if is_variant is not None:\n",
    "        is_variant = is_variant.view('u1')\n",
    "    out = opt_is_cas9_target(seq, revcomp=revcomp, is_variant=is_variant)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0461b858-9416-45cb-9567-061224b4169d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_cds_idx(chrom):\n",
    "    \"\"\"Locate CDS features on the reference sequence.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=int)\n",
    "    chrom = f'AfunGA1_{chrom}'\n",
    "    for i, (seqid, start, end) in enumerate(cdss_chroms[['seqid', 'start', 'end']]):\n",
    "        if seqid == chrom:\n",
    "            out[start-1:end] = i + 1  # start indexing from 1\n",
    "    return out\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f81dcc3-8ca7-4ae6-8b40-4e0c186fdf9d",
   "metadata": {},
   "source": [
    "This gives each CDS a unique index -- however, if there are ovelapping CDS, the last listed one counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e4e1cf85-fe3a-4349-950b-0a97c9d63d85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_gene_idx(chrom):\n",
    "    \"\"\"Locate protein-coding genes on the reference sequence.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=int)\n",
    "    chrom = f'AfunGA1_{chrom}'\n",
    "    for i, (seqid, start, end) in enumerate(genes_coding_chroms[['seqid', 'start', 'end']]):\n",
    "        if seqid == chrom:\n",
    "            out[start-1:end] = i + 1  # start indexing from 1\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aef75df5-45e3-4890-b7de-f6de70d5b254",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_target_in_cds(chrom):\n",
    "    \"\"\"Locate targets within a single CDS.\"\"\"\n",
    "    t = get_cds_idx(chrom)\n",
    "    out = opt_all_subsequent(t, 21) > 0\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "242ba5cb-ca03-4041-9e12-540b2f69736d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#@functools.lru_cache(maxsize=None)\n",
    "def get_is_variant_individual(chrom, sample_idx):\n",
    "    \"\"\"Locate variant sites for a given sample.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=bool)\n",
    "    pos = af1.snp_calls(chrom).variant_position.values\n",
    "    sample_idx_list = [int(sample_idx)]\n",
    "    g = af1.snp_calls(chrom, sample_indices = sample_idx_list).call_genotype\n",
    "    loc = np.any(g > 0, axis=(1,2))\n",
    "    loc = loc.compute().values\n",
    "    idx = pos[loc] - 1\n",
    "    out[idx] = True\n",
    "\n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd59838f-9636-4297-a709-9da7ca97c213",
   "metadata": {},
   "source": [
    "# Set up cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "45a72b46-3efa-4451-b9fd-3ab3ed92c777",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# gateway = Gateway()\n",
    "# for cl in gateway.list_clusters():\n",
    "#     gateway.connect(cl.name).shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ba6228b3-54a9-4fb4-8e46-fe062b66433f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0f60829c82ff478d8a1cbe4dd0472f19",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HTML(value='<h2>GatewayCluster</h2>'), HBox(children=(HTML(value='\\n<div>\\n<style scoped>\\n    …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gateway = Gateway()\n",
    "conda_prefix = os.environ[\"CONDA_PREFIX\"]\n",
    "current_environment = 'global/'+conda_prefix.split('/')[5]\n",
    "cluster = gateway.new_cluster(\n",
    "    profile='standard', \n",
    "    conda_environment = current_environment,\n",
    ")\n",
    "cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f7b69903-d74a-48fc-91b5-de5893ef8392",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-08 06:25:16,961 - distributed.client - ERROR - Failed to reconnect to scheduler after 30.00 seconds, closing client\n"
     ]
    }
   ],
   "source": [
    "client=cluster.get_client()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "268019bf-188c-4eb0-a905-e09f58e807ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cluster.scale(60)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fbd79e1-81b7-4027-b9c1-946c6738d500",
   "metadata": {},
   "source": [
    "# Run analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fde3d6e4-7ac3-4353-9599-75a43c1de752",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def identify_targets_population_cumulative(chrom, sample_idx):\n",
    "\n",
    "    #Find targets in cds (not using site filter)\n",
    "    base_loc = get_target_in_cds(chrom) \n",
    "    loc = base_loc & get_is_cas9_target(chrom, revcomp=True)\n",
    "    \n",
    "    #index genes\n",
    "    gene_idx = get_gene_idx(chrom)\n",
    "    \n",
    "    # setup outputs\n",
    "    out_sites = []\n",
    "    out_targets = []\n",
    "    out_genes = []\n",
    "    \n",
    "    # setup variation\n",
    "    is_variant = np.zeros(gene_idx.shape[0], dtype=bool)\n",
    "    \n",
    "    #initial numbers\n",
    "    out_sites.append(0)\n",
    "    \n",
    "    n_targets = np.count_nonzero(loc)\n",
    "    out_targets.append(n_targets)\n",
    "    \n",
    "    n_genes = len(np.unique(gene_idx[loc]))\n",
    "    out_genes.append(n_genes)\n",
    "    \n",
    "    #randomise order of samples\n",
    "    rng = np.random.default_rng()\n",
    "    sample_idx = rng.choice(sample_idx, size=len(sample_idx),\n",
    "                              replace=False)\n",
    "    \n",
    "    for i, sidx in enumerate(sample_idx):\n",
    "        \n",
    "        #accumulate variation\n",
    "        is_variant |= get_is_variant_individual(chrom, sidx)\n",
    "        \n",
    "        # locate sites\n",
    "        n_sites = np.count_nonzero(is_variant)\n",
    "        out_sites.append(n_sites)\n",
    "        \n",
    "        # locate targets and genes\n",
    "        loc = base_loc & get_is_cas9_target(chrom, revcomp=True, is_variant=is_variant)\n",
    "        n_targets = np.count_nonzero(loc)\n",
    "        out_targets.append(n_targets)\n",
    "\n",
    "        n_genes = len(np.unique(gene_idx[loc]))\n",
    "        out_genes.append(n_genes)\n",
    "        if i%10 == 0:\n",
    "            print(f'Sample {i} done')\n",
    "\n",
    "\n",
    "    return np.array([out_sites, out_targets, out_genes]).T\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "44475a5b-b2a3-44b1-ab88-74923b9d69df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_analysis(chrom, pop, sample_idx):\n",
    "    \n",
    "    outdir = f'cumulative_results/{pop}/'\n",
    "    if os.path.exists(f'{outdir}/target_info_{chrom}.npy'): \n",
    "        print(f'{outdir}/target_info_{chrom}.npy already exists, skipping computation')\n",
    "    else:\n",
    "        if not os.path.isdir(outdir):\n",
    "            !mkdir {outdir}\n",
    "        \n",
    "        print(f'Computing for {len(sample_idx)} samples in cohort {pop}')\n",
    "        target_info = identify_targets_population_cumulative(chrom, sample_idx)\n",
    "        np.save(f'{outdir}/target_info_{chrom}.npy', target_info)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d3a8d8bb-e546-41c1-aaa3-9d89a9ab9535",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    }
   ],
   "source": [
    "af1.add_extra_metadata(pd.read_csv('../../metadata/supp1.csv'))\n",
    "meta = af1.sample_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "603c50ad-67c4-41dc-99fb-ebff25924624",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sample_idx_dict = dict({'hicov': tuple(meta.loc[meta.subset_2=='Y'].index.values)})\n",
    "for pop in meta.geographic_cohort.unique():\n",
    "    sample_idx = meta.loc[(meta.geographic_cohort==pop) & (meta.subset_2=='Y')].index.values\n",
    "    sample_idx_dict[pop] = tuple(sample_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a19d58df-7b54-480c-aa6d-a9ee39cffed4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'hicov': (0,\n",
       "  1,\n",
       "  2,\n",
       "  3,\n",
       "  4,\n",
       "  5,\n",
       "  6,\n",
       "  7,\n",
       "  8,\n",
       "  9,\n",
       "  10,\n",
       "  11,\n",
       "  13,\n",
       "  14,\n",
       "  15,\n",
       "  18,\n",
       "  19,\n",
       "  20,\n",
       "  21,\n",
       "  23,\n",
       "  24,\n",
       "  25,\n",
       "  26,\n",
       "  27,\n",
       "  28,\n",
       "  29,\n",
       "  30,\n",
       "  31,\n",
       "  32,\n",
       "  33,\n",
       "  34,\n",
       "  35,\n",
       "  36,\n",
       "  37,\n",
       "  38,\n",
       "  39,\n",
       "  40,\n",
       "  41,\n",
       "  42,\n",
       "  43,\n",
       "  44,\n",
       "  45,\n",
       "  46,\n",
       "  47,\n",
       "  48,\n",
       "  49,\n",
       "  50,\n",
       "  51,\n",
       "  52,\n",
       "  53,\n",
       "  54,\n",
       "  55,\n",
       "  56,\n",
       "  57,\n",
       "  58,\n",
       "  59,\n",
       "  60,\n",
       "  61,\n",
       "  62,\n",
       "  63,\n",
       "  64,\n",
       "  65,\n",
       "  66,\n",
       "  67,\n",
       "  68,\n",
       "  69,\n",
       "  70,\n",
       "  71,\n",
       "  72,\n",
       "  73,\n",
       "  74,\n",
       "  75,\n",
       "  76,\n",
       "  77,\n",
       "  78,\n",
       "  79,\n",
       "  80,\n",
       "  81,\n",
       "  82,\n",
       "  83,\n",
       "  84,\n",
       "  85,\n",
       "  86,\n",
       "  87,\n",
       "  88,\n",
       "  89,\n",
       "  90,\n",
       "  91,\n",
       "  92,\n",
       "  93,\n",
       "  94,\n",
       "  95,\n",
       "  96,\n",
       "  97,\n",
       "  98,\n",
       "  99,\n",
       "  100,\n",
       "  101,\n",
       "  102,\n",
       "  103,\n",
       "  104,\n",
       "  105,\n",
       "  106,\n",
       "  107,\n",
       "  108,\n",
       "  109,\n",
       "  110,\n",
       "  111,\n",
       "  112,\n",
       "  113,\n",
       "  114,\n",
       "  115,\n",
       "  116,\n",
       "  117,\n",
       "  118,\n",
       "  119,\n",
       "  120,\n",
       "  121,\n",
       "  122,\n",
       "  123,\n",
       "  124,\n",
       "  125,\n",
       "  126,\n",
       "  127,\n",
       "  128,\n",
       "  129,\n",
       "  130,\n",
       "  131,\n",
       "  132,\n",
       "  133,\n",
       "  134,\n",
       "  135,\n",
       "  136,\n",
       "  137,\n",
       "  138,\n",
       "  139,\n",
       "  140,\n",
       "  141,\n",
       "  142,\n",
       "  143,\n",
       "  144,\n",
       "  145,\n",
       "  146,\n",
       "  147,\n",
       "  148,\n",
       "  149,\n",
       "  150,\n",
       "  151,\n",
       "  152,\n",
       "  153,\n",
       "  154,\n",
       "  155,\n",
       "  156,\n",
       "  157,\n",
       "  159,\n",
       "  160,\n",
       "  161,\n",
       "  162,\n",
       "  163,\n",
       "  164,\n",
       "  165,\n",
       "  166,\n",
       "  167,\n",
       "  169,\n",
       "  170,\n",
       "  171,\n",
       "  172,\n",
       "  173,\n",
       "  174,\n",
       "  175,\n",
       "  176,\n",
       "  177,\n",
       "  178,\n",
       "  179,\n",
       "  180,\n",
       "  181,\n",
       "  182,\n",
       "  183,\n",
       "  184,\n",
       "  185,\n",
       "  187,\n",
       "  188,\n",
       "  189,\n",
       "  190,\n",
       "  191,\n",
       "  192,\n",
       "  193,\n",
       "  194,\n",
       "  195,\n",
       "  196,\n",
       "  197,\n",
       "  199,\n",
       "  200,\n",
       "  201,\n",
       "  202,\n",
       "  203,\n",
       "  204,\n",
       "  205,\n",
       "  206,\n",
       "  207,\n",
       "  208,\n",
       "  209,\n",
       "  210,\n",
       "  211,\n",
       "  212,\n",
       "  213,\n",
       "  214,\n",
       "  215,\n",
       "  216,\n",
       "  217,\n",
       "  218,\n",
       "  219,\n",
       "  220,\n",
       "  221,\n",
       "  222,\n",
       "  223,\n",
       "  224,\n",
       "  225,\n",
       "  226,\n",
       "  227,\n",
       "  228,\n",
       "  229,\n",
       "  230,\n",
       "  231,\n",
       "  232,\n",
       "  233,\n",
       "  234,\n",
       "  235,\n",
       "  236,\n",
       "  237,\n",
       "  238,\n",
       "  239,\n",
       "  240,\n",
       "  241,\n",
       "  242,\n",
       "  243,\n",
       "  244,\n",
       "  245,\n",
       "  246,\n",
       "  247,\n",
       "  248,\n",
       "  249,\n",
       "  250,\n",
       "  251,\n",
       "  252,\n",
       "  253,\n",
       "  254,\n",
       "  255,\n",
       "  256,\n",
       "  257,\n",
       "  258,\n",
       "  259,\n",
       "  260,\n",
       "  261,\n",
       "  262,\n",
       "  263,\n",
       "  264,\n",
       "  265,\n",
       "  266,\n",
       "  267,\n",
       "  268,\n",
       "  269,\n",
       "  270,\n",
       "  271,\n",
       "  272,\n",
       "  273,\n",
       "  274,\n",
       "  275,\n",
       "  276,\n",
       "  277,\n",
       "  278,\n",
       "  279,\n",
       "  280,\n",
       "  281,\n",
       "  282,\n",
       "  283,\n",
       "  285,\n",
       "  287,\n",
       "  288,\n",
       "  289,\n",
       "  290,\n",
       "  291,\n",
       "  292,\n",
       "  293,\n",
       "  294,\n",
       "  296,\n",
       "  297,\n",
       "  298,\n",
       "  299,\n",
       "  300,\n",
       "  301,\n",
       "  302,\n",
       "  303,\n",
       "  304,\n",
       "  305,\n",
       "  306,\n",
       "  307,\n",
       "  308,\n",
       "  309,\n",
       "  310,\n",
       "  311,\n",
       "  312,\n",
       "  313,\n",
       "  314,\n",
       "  316,\n",
       "  317,\n",
       "  318,\n",
       "  319,\n",
       "  320,\n",
       "  321,\n",
       "  322,\n",
       "  323,\n",
       "  324,\n",
       "  325,\n",
       "  326,\n",
       "  327,\n",
       "  328,\n",
       "  329,\n",
       "  330,\n",
       "  331,\n",
       "  332,\n",
       "  333,\n",
       "  334,\n",
       "  335,\n",
       "  336,\n",
       "  337,\n",
       "  338,\n",
       "  339,\n",
       "  340,\n",
       "  341,\n",
       "  342,\n",
       "  343,\n",
       "  344,\n",
       "  345,\n",
       "  346,\n",
       "  347,\n",
       "  348,\n",
       "  349,\n",
       "  350,\n",
       "  351,\n",
       "  352,\n",
       "  353,\n",
       "  354,\n",
       "  355,\n",
       "  356,\n",
       "  357,\n",
       "  358,\n",
       "  359,\n",
       "  360,\n",
       "  361,\n",
       "  362,\n",
       "  363,\n",
       "  365,\n",
       "  368,\n",
       "  369,\n",
       "  370,\n",
       "  371,\n",
       "  372,\n",
       "  373,\n",
       "  374,\n",
       "  375,\n",
       "  376,\n",
       "  377,\n",
       "  378,\n",
       "  379,\n",
       "  380,\n",
       "  381,\n",
       "  382,\n",
       "  383,\n",
       "  384,\n",
       "  385,\n",
       "  386,\n",
       "  387,\n",
       "  388,\n",
       "  389,\n",
       "  390,\n",
       "  391,\n",
       "  392,\n",
       "  393,\n",
       "  394,\n",
       "  395,\n",
       "  396,\n",
       "  397,\n",
       "  398,\n",
       "  399,\n",
       "  400,\n",
       "  401,\n",
       "  402,\n",
       "  403,\n",
       "  404,\n",
       "  405,\n",
       "  406,\n",
       "  407,\n",
       "  410,\n",
       "  411,\n",
       "  412,\n",
       "  413,\n",
       "  414,\n",
       "  415,\n",
       "  417,\n",
       "  418,\n",
       "  420,\n",
       "  421,\n",
       "  422,\n",
       "  423,\n",
       "  424,\n",
       "  425,\n",
       "  426,\n",
       "  428,\n",
       "  430,\n",
       "  431,\n",
       "  432,\n",
       "  435,\n",
       "  436,\n",
       "  440,\n",
       "  441,\n",
       "  442,\n",
       "  443,\n",
       "  444,\n",
       "  445,\n",
       "  446,\n",
       "  447,\n",
       "  448,\n",
       "  449,\n",
       "  450,\n",
       "  451,\n",
       "  452,\n",
       "  453,\n",
       "  454,\n",
       "  455,\n",
       "  456,\n",
       "  457,\n",
       "  458,\n",
       "  459,\n",
       "  460,\n",
       "  461,\n",
       "  462,\n",
       "  463,\n",
       "  464,\n",
       "  465,\n",
       "  466,\n",
       "  467,\n",
       "  468,\n",
       "  469,\n",
       "  470,\n",
       "  471,\n",
       "  472,\n",
       "  473,\n",
       "  475,\n",
       "  476,\n",
       "  477,\n",
       "  478,\n",
       "  479,\n",
       "  480,\n",
       "  481,\n",
       "  482,\n",
       "  483,\n",
       "  484,\n",
       "  485,\n",
       "  486,\n",
       "  493,\n",
       "  494,\n",
       "  496,\n",
       "  497,\n",
       "  498,\n",
       "  499,\n",
       "  500,\n",
       "  501,\n",
       "  502,\n",
       "  503,\n",
       "  504,\n",
       "  505,\n",
       "  506,\n",
       "  507,\n",
       "  508,\n",
       "  509,\n",
       "  510,\n",
       "  511,\n",
       "  512,\n",
       "  513,\n",
       "  514,\n",
       "  515,\n",
       "  516,\n",
       "  517,\n",
       "  518,\n",
       "  519,\n",
       "  520,\n",
       "  521,\n",
       "  522,\n",
       "  523,\n",
       "  524,\n",
       "  525,\n",
       "  526,\n",
       "  527,\n",
       "  528,\n",
       "  529,\n",
       "  530,\n",
       "  531,\n",
       "  532,\n",
       "  533,\n",
       "  534,\n",
       "  535,\n",
       "  536,\n",
       "  537,\n",
       "  538,\n",
       "  539,\n",
       "  540,\n",
       "  541,\n",
       "  542,\n",
       "  543,\n",
       "  544,\n",
       "  545,\n",
       "  546,\n",
       "  547,\n",
       "  548,\n",
       "  549,\n",
       "  550,\n",
       "  551,\n",
       "  552,\n",
       "  553,\n",
       "  554,\n",
       "  555,\n",
       "  556,\n",
       "  557,\n",
       "  558,\n",
       "  559,\n",
       "  560,\n",
       "  561,\n",
       "  562,\n",
       "  563,\n",
       "  564,\n",
       "  565,\n",
       "  566,\n",
       "  567,\n",
       "  571,\n",
       "  572,\n",
       "  573,\n",
       "  574,\n",
       "  575,\n",
       "  576,\n",
       "  577,\n",
       "  578,\n",
       "  579,\n",
       "  580,\n",
       "  581,\n",
       "  582,\n",
       "  583,\n",
       "  584,\n",
       "  585,\n",
       "  586,\n",
       "  587,\n",
       "  588,\n",
       "  589,\n",
       "  590,\n",
       "  591,\n",
       "  592,\n",
       "  593,\n",
       "  594,\n",
       "  595,\n",
       "  596,\n",
       "  597,\n",
       "  598,\n",
       "  599,\n",
       "  600,\n",
       "  601,\n",
       "  602,\n",
       "  603,\n",
       "  604,\n",
       "  605,\n",
       "  606,\n",
       "  607,\n",
       "  608,\n",
       "  609,\n",
       "  610,\n",
       "  611,\n",
       "  612,\n",
       "  613,\n",
       "  614,\n",
       "  615,\n",
       "  616,\n",
       "  617,\n",
       "  618,\n",
       "  619,\n",
       "  620,\n",
       "  621,\n",
       "  622,\n",
       "  623,\n",
       "  624,\n",
       "  625,\n",
       "  626,\n",
       "  627,\n",
       "  628,\n",
       "  629,\n",
       "  630,\n",
       "  631,\n",
       "  632,\n",
       "  633,\n",
       "  634,\n",
       "  635,\n",
       "  636,\n",
       "  637,\n",
       "  638,\n",
       "  639,\n",
       "  640,\n",
       "  641,\n",
       "  642,\n",
       "  643,\n",
       "  644,\n",
       "  645,\n",
       "  646,\n",
       "  647,\n",
       "  648,\n",
       "  649,\n",
       "  650,\n",
       "  651,\n",
       "  652,\n",
       "  653,\n",
       "  654,\n",
       "  655),\n",
       " 'Ghana_Northern-Region': (0,\n",
       "  1,\n",
       "  2,\n",
       "  3,\n",
       "  4,\n",
       "  5,\n",
       "  6,\n",
       "  7,\n",
       "  8,\n",
       "  9,\n",
       "  10,\n",
       "  11,\n",
       "  13,\n",
       "  14,\n",
       "  15,\n",
       "  18,\n",
       "  19,\n",
       "  20,\n",
       "  21,\n",
       "  23,\n",
       "  24,\n",
       "  25,\n",
       "  26,\n",
       "  27,\n",
       "  28,\n",
       "  29,\n",
       "  30,\n",
       "  31,\n",
       "  32,\n",
       "  33,\n",
       "  34,\n",
       "  35),\n",
       " 'Gabon_Haut-Ogooue': (36,\n",
       "  37,\n",
       "  38,\n",
       "  39,\n",
       "  40,\n",
       "  41,\n",
       "  42,\n",
       "  43,\n",
       "  44,\n",
       "  45,\n",
       "  46,\n",
       "  47,\n",
       "  48,\n",
       "  49,\n",
       "  50,\n",
       "  51,\n",
       "  52,\n",
       "  53,\n",
       "  54,\n",
       "  55,\n",
       "  56,\n",
       "  57,\n",
       "  58,\n",
       "  59,\n",
       "  60,\n",
       "  61,\n",
       "  62,\n",
       "  63,\n",
       "  64,\n",
       "  65,\n",
       "  66,\n",
       "  67,\n",
       "  68,\n",
       "  69,\n",
       "  70,\n",
       "  71,\n",
       "  72,\n",
       "  73,\n",
       "  74,\n",
       "  75),\n",
       " 'CAR_Ombella-MPoko': (76, 77, 78, 79, 80, 81, 82, 83, 84, 85),\n",
       " 'Cameroon_Adamawa': (86,\n",
       "  87,\n",
       "  88,\n",
       "  89,\n",
       "  90,\n",
       "  91,\n",
       "  92,\n",
       "  93,\n",
       "  94,\n",
       "  95,\n",
       "  96,\n",
       "  97,\n",
       "  98,\n",
       "  99,\n",
       "  100,\n",
       "  101,\n",
       "  102,\n",
       "  103,\n",
       "  104,\n",
       "  105,\n",
       "  106,\n",
       "  107,\n",
       "  108,\n",
       "  109,\n",
       "  110,\n",
       "  111,\n",
       "  112,\n",
       "  113,\n",
       "  114,\n",
       "  115,\n",
       "  116,\n",
       "  117,\n",
       "  118,\n",
       "  119,\n",
       "  120,\n",
       "  121,\n",
       "  122,\n",
       "  123,\n",
       "  124,\n",
       "  125,\n",
       "  126,\n",
       "  127,\n",
       "  128,\n",
       "  129,\n",
       "  130),\n",
       " 'Ghana_Ashanti-Region': (131,\n",
       "  132,\n",
       "  133,\n",
       "  134,\n",
       "  135,\n",
       "  136,\n",
       "  137,\n",
       "  138,\n",
       "  139,\n",
       "  140,\n",
       "  141,\n",
       "  142,\n",
       "  143,\n",
       "  144,\n",
       "  145,\n",
       "  146,\n",
       "  147,\n",
       "  148,\n",
       "  149,\n",
       "  150,\n",
       "  151,\n",
       "  152,\n",
       "  153,\n",
       "  154,\n",
       "  155,\n",
       "  156,\n",
       "  157,\n",
       "  159,\n",
       "  160,\n",
       "  161),\n",
       " 'Malawi_Southern-Region': (162,\n",
       "  163,\n",
       "  164,\n",
       "  165,\n",
       "  166,\n",
       "  167,\n",
       "  169,\n",
       "  170,\n",
       "  171,\n",
       "  172,\n",
       "  173,\n",
       "  174,\n",
       "  175,\n",
       "  176,\n",
       "  177,\n",
       "  178,\n",
       "  179),\n",
       " 'Mozambique_Maputo': (180,\n",
       "  181,\n",
       "  182,\n",
       "  183,\n",
       "  184,\n",
       "  185,\n",
       "  187,\n",
       "  188,\n",
       "  189,\n",
       "  190,\n",
       "  191,\n",
       "  192,\n",
       "  193,\n",
       "  194,\n",
       "  195,\n",
       "  196,\n",
       "  197,\n",
       "  199,\n",
       "  200,\n",
       "  201,\n",
       "  493,\n",
       "  494,\n",
       "  496,\n",
       "  497,\n",
       "  498,\n",
       "  499,\n",
       "  500,\n",
       "  501,\n",
       "  502,\n",
       "  503,\n",
       "  504,\n",
       "  505,\n",
       "  506,\n",
       "  507,\n",
       "  508,\n",
       "  509,\n",
       "  510,\n",
       "  511,\n",
       "  512,\n",
       "  513,\n",
       "  514,\n",
       "  515,\n",
       "  516,\n",
       "  517,\n",
       "  518,\n",
       "  519,\n",
       "  520,\n",
       "  521,\n",
       "  522,\n",
       "  523,\n",
       "  524,\n",
       "  525,\n",
       "  526,\n",
       "  527,\n",
       "  528,\n",
       "  529,\n",
       "  530,\n",
       "  531,\n",
       "  532,\n",
       "  533,\n",
       "  534,\n",
       "  535,\n",
       "  536,\n",
       "  537,\n",
       "  538,\n",
       "  539,\n",
       "  540,\n",
       "  541,\n",
       "  542,\n",
       "  543,\n",
       "  544,\n",
       "  545,\n",
       "  546,\n",
       "  547,\n",
       "  548,\n",
       "  549,\n",
       "  550,\n",
       "  551,\n",
       "  552,\n",
       "  553,\n",
       "  554,\n",
       "  555,\n",
       "  556,\n",
       "  557,\n",
       "  558,\n",
       "  559,\n",
       "  560,\n",
       "  561,\n",
       "  562),\n",
       " 'Uganda_Eastern-Region': (202,\n",
       "  203,\n",
       "  204,\n",
       "  205,\n",
       "  206,\n",
       "  207,\n",
       "  208,\n",
       "  209,\n",
       "  210,\n",
       "  211,\n",
       "  212,\n",
       "  213,\n",
       "  214,\n",
       "  215,\n",
       "  216,\n",
       "  217,\n",
       "  218,\n",
       "  219,\n",
       "  220,\n",
       "  221,\n",
       "  222,\n",
       "  223,\n",
       "  224,\n",
       "  225,\n",
       "  226,\n",
       "  227,\n",
       "  228,\n",
       "  229,\n",
       "  230,\n",
       "  231,\n",
       "  232,\n",
       "  233,\n",
       "  234,\n",
       "  235,\n",
       "  236,\n",
       "  237,\n",
       "  238,\n",
       "  239,\n",
       "  240,\n",
       "  241,\n",
       "  242,\n",
       "  243,\n",
       "  244,\n",
       "  245,\n",
       "  246,\n",
       "  247,\n",
       "  248,\n",
       "  249,\n",
       "  250),\n",
       " 'Benin_Atlantique-Dept': (251,\n",
       "  252,\n",
       "  253,\n",
       "  254,\n",
       "  255,\n",
       "  256,\n",
       "  257,\n",
       "  258,\n",
       "  259,\n",
       "  260,\n",
       "  261,\n",
       "  262,\n",
       "  263,\n",
       "  264,\n",
       "  265,\n",
       "  266,\n",
       "  267,\n",
       "  268,\n",
       "  269,\n",
       "  270,\n",
       "  271,\n",
       "  272,\n",
       "  273,\n",
       "  274,\n",
       "  275,\n",
       "  276,\n",
       "  277,\n",
       "  278,\n",
       "  279,\n",
       "  280,\n",
       "  281,\n",
       "  282,\n",
       "  283,\n",
       "  285,\n",
       "  287),\n",
       " 'DRC_Kinshasa': (288,\n",
       "  289,\n",
       "  290,\n",
       "  291,\n",
       "  292,\n",
       "  293,\n",
       "  294,\n",
       "  296,\n",
       "  297,\n",
       "  298,\n",
       "  299,\n",
       "  300,\n",
       "  301,\n",
       "  302,\n",
       "  303,\n",
       "  304,\n",
       "  305,\n",
       "  306,\n",
       "  307,\n",
       "  308,\n",
       "  309,\n",
       "  310,\n",
       "  311,\n",
       "  312,\n",
       "  313,\n",
       "  314,\n",
       "  316,\n",
       "  317,\n",
       "  318,\n",
       "  319,\n",
       "  320,\n",
       "  321),\n",
       " 'Nigeria_Ogun-State': (322,\n",
       "  323,\n",
       "  324,\n",
       "  325,\n",
       "  326,\n",
       "  327,\n",
       "  328,\n",
       "  329,\n",
       "  330,\n",
       "  331,\n",
       "  332,\n",
       "  333,\n",
       "  334,\n",
       "  335,\n",
       "  336,\n",
       "  337,\n",
       "  338,\n",
       "  339,\n",
       "  340,\n",
       "  341,\n",
       "  342,\n",
       "  343,\n",
       "  344,\n",
       "  345,\n",
       "  346,\n",
       "  347,\n",
       "  348,\n",
       "  349,\n",
       "  350,\n",
       "  351,\n",
       "  352,\n",
       "  353,\n",
       "  354,\n",
       "  355,\n",
       "  356,\n",
       "  357,\n",
       "  358,\n",
       "  359,\n",
       "  360,\n",
       "  361,\n",
       "  362),\n",
       " 'Zambia_Eastern-Prov': (363,\n",
       "  365,\n",
       "  368,\n",
       "  369,\n",
       "  370,\n",
       "  371,\n",
       "  372,\n",
       "  373,\n",
       "  374,\n",
       "  375,\n",
       "  376,\n",
       "  377,\n",
       "  378,\n",
       "  379,\n",
       "  380,\n",
       "  381,\n",
       "  382,\n",
       "  383,\n",
       "  384,\n",
       "  385,\n",
       "  386,\n",
       "  387,\n",
       "  388,\n",
       "  389,\n",
       "  390,\n",
       "  391,\n",
       "  392,\n",
       "  393,\n",
       "  394,\n",
       "  395,\n",
       "  396,\n",
       "  397,\n",
       "  398,\n",
       "  399,\n",
       "  400,\n",
       "  401,\n",
       "  402,\n",
       "  403,\n",
       "  404,\n",
       "  405),\n",
       " 'Kenya_Nyanza-Prov': (406,\n",
       "  407,\n",
       "  410,\n",
       "  411,\n",
       "  412,\n",
       "  413,\n",
       "  414,\n",
       "  415,\n",
       "  417,\n",
       "  418,\n",
       "  420,\n",
       "  421,\n",
       "  422,\n",
       "  423,\n",
       "  424,\n",
       "  425,\n",
       "  426,\n",
       "  428,\n",
       "  430,\n",
       "  431,\n",
       "  432,\n",
       "  435,\n",
       "  436,\n",
       "  440,\n",
       "  441,\n",
       "  442,\n",
       "  443,\n",
       "  445,\n",
       "  446,\n",
       "  447,\n",
       "  448,\n",
       "  449,\n",
       "  450,\n",
       "  451,\n",
       "  453,\n",
       "  454,\n",
       "  455,\n",
       "  456,\n",
       "  457,\n",
       "  458,\n",
       "  459,\n",
       "  460,\n",
       "  461,\n",
       "  462,\n",
       "  482,\n",
       "  483,\n",
       "  484,\n",
       "  485,\n",
       "  486),\n",
       " 'Kenya_Western-Prov': (444,\n",
       "  452,\n",
       "  463,\n",
       "  464,\n",
       "  465,\n",
       "  466,\n",
       "  467,\n",
       "  468,\n",
       "  469,\n",
       "  470,\n",
       "  471,\n",
       "  472,\n",
       "  473,\n",
       "  475,\n",
       "  476,\n",
       "  477,\n",
       "  478,\n",
       "  479,\n",
       "  480,\n",
       "  481),\n",
       " 'Tanzania_Morogoro-Region': (563, 564, 565, 566, 567, 571, 572),\n",
       " 'DRC_Haut-Uele': (573,\n",
       "  574,\n",
       "  575,\n",
       "  576,\n",
       "  577,\n",
       "  578,\n",
       "  579,\n",
       "  580,\n",
       "  581,\n",
       "  582,\n",
       "  583,\n",
       "  584,\n",
       "  585,\n",
       "  586,\n",
       "  587,\n",
       "  588,\n",
       "  589,\n",
       "  590,\n",
       "  591,\n",
       "  592,\n",
       "  593,\n",
       "  594,\n",
       "  595,\n",
       "  596,\n",
       "  597,\n",
       "  598,\n",
       "  599,\n",
       "  600,\n",
       "  601,\n",
       "  602,\n",
       "  603,\n",
       "  604,\n",
       "  605,\n",
       "  606,\n",
       "  607,\n",
       "  608,\n",
       "  609,\n",
       "  610,\n",
       "  611,\n",
       "  612,\n",
       "  613,\n",
       "  614,\n",
       "  615),\n",
       " 'Mozambique_Cabo-Delgado': (616,\n",
       "  617,\n",
       "  618,\n",
       "  619,\n",
       "  620,\n",
       "  621,\n",
       "  622,\n",
       "  623,\n",
       "  624,\n",
       "  625,\n",
       "  626,\n",
       "  627,\n",
       "  628,\n",
       "  629,\n",
       "  630,\n",
       "  631,\n",
       "  632,\n",
       "  633,\n",
       "  634,\n",
       "  635,\n",
       "  636,\n",
       "  637,\n",
       "  638,\n",
       "  639,\n",
       "  640,\n",
       "  641,\n",
       "  642,\n",
       "  643,\n",
       "  644,\n",
       "  645,\n",
       "  646,\n",
       "  647,\n",
       "  648,\n",
       "  649,\n",
       "  650,\n",
       "  651,\n",
       "  652,\n",
       "  653,\n",
       "  654,\n",
       "  655)}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_idx_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b5c70937-5192-43c3-9307-0588a773ae2b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cumulative_results/Ghana_Northern-Region//target_info_X.npy already exists, skipping computation\n"
     ]
    }
   ],
   "source": [
    "#First test on a small set\n",
    "run_analysis('X', 'Ghana_Northern-Region', sample_idx = sample_idx_dict['Ghana_Northern-Region'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1adb0896-359f-44a5-8705-6f5353d74ac2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cumulative_results/Ghana_Northern-Region//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Ghana_Northern-Region on chrom X\n",
      "cumulative_results/Gabon_Haut-Ogooue//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Gabon_Haut-Ogooue on chrom X\n",
      "cumulative_results/CAR_Ombella-MPoko//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort CAR_Ombella-MPoko on chrom X\n",
      "cumulative_results/Cameroon_Adamawa//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Cameroon_Adamawa on chrom X\n",
      "cumulative_results/Ghana_Ashanti-Region//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Ghana_Ashanti-Region on chrom X\n",
      "cumulative_results/Malawi_Southern-Region//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Malawi_Southern-Region on chrom X\n",
      "cumulative_results/Mozambique_Maputo//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Mozambique_Maputo on chrom X\n",
      "cumulative_results/Uganda_Eastern-Region//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Uganda_Eastern-Region on chrom X\n",
      "cumulative_results/Benin_Atlantique-Dept//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Benin_Atlantique-Dept on chrom X\n",
      "cumulative_results/DRC_Kinshasa//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort DRC_Kinshasa on chrom X\n",
      "cumulative_results/Nigeria_Ogun-State//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Nigeria_Ogun-State on chrom X\n",
      "cumulative_results/Zambia_Eastern-Prov//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Zambia_Eastern-Prov on chrom X\n",
      "cumulative_results/Kenya_Nyanza-Prov//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Kenya_Nyanza-Prov on chrom X\n",
      "cumulative_results/Kenya_Western-Prov//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Kenya_Western-Prov on chrom X\n",
      "cumulative_results/Tanzania_Morogoro-Region//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Tanzania_Morogoro-Region on chrom X\n",
      "cumulative_results/DRC_Haut-Uele//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort DRC_Haut-Uele on chrom X\n",
      "cumulative_results/Mozambique_Cabo-Delgado//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort Mozambique_Cabo-Delgado on chrom X\n",
      "Computing for 32 samples in cohort Ghana_Northern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Ghana_Northern-Region on chrom 3RL\n",
      "Computing for 40 samples in cohort Gabon_Haut-Ogooue\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Gabon_Haut-Ogooue on chrom 3RL\n",
      "Computing for 10 samples in cohort CAR_Ombella-MPoko\n",
      "Sample 0 done                    \n",
      "Done for cohort CAR_Ombella-MPoko on chrom 3RL\n",
      "Computing for 45 samples in cohort Cameroon_Adamawa\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Cameroon_Adamawa on chrom 3RL\n",
      "Computing for 30 samples in cohort Ghana_Ashanti-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Done for cohort Ghana_Ashanti-Region on chrom 3RL\n",
      "Computing for 17 samples in cohort Malawi_Southern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Done for cohort Malawi_Southern-Region on chrom 3RL\n",
      "Computing for 89 samples in cohort Mozambique_Maputo\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Sample 50 done                   \n",
      "Sample 60 done                   \n",
      "Sample 70 done                   \n",
      "Sample 80 done                   \n",
      "Done for cohort Mozambique_Maputo on chrom 3RL\n",
      "Computing for 49 samples in cohort Uganda_Eastern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Uganda_Eastern-Region on chrom 3RL\n",
      "Computing for 35 samples in cohort Benin_Atlantique-Dept\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Benin_Atlantique-Dept on chrom 3RL\n",
      "Computing for 32 samples in cohort DRC_Kinshasa\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort DRC_Kinshasa on chrom 3RL\n",
      "Computing for 41 samples in cohort Nigeria_Ogun-State\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Nigeria_Ogun-State on chrom 3RL\n",
      "Computing for 40 samples in cohort Zambia_Eastern-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Zambia_Eastern-Prov on chrom 3RL\n",
      "Computing for 49 samples in cohort Kenya_Nyanza-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Kenya_Nyanza-Prov on chrom 3RL\n",
      "Computing for 20 samples in cohort Kenya_Western-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Done for cohort Kenya_Western-Prov on chrom 3RL\n",
      "Computing for 7 samples in cohort Tanzania_Morogoro-Region\n",
      "Sample 0 done                    \n",
      "Done for cohort Tanzania_Morogoro-Region on chrom 3RL\n",
      "Computing for 43 samples in cohort DRC_Haut-Uele\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort DRC_Haut-Uele on chrom 3RL\n",
      "Computing for 40 samples in cohort Mozambique_Cabo-Delgado\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Mozambique_Cabo-Delgado on chrom 3RL\n",
      "Computing for 32 samples in cohort Ghana_Northern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Ghana_Northern-Region on chrom 2RL\n",
      "Computing for 40 samples in cohort Gabon_Haut-Ogooue\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Gabon_Haut-Ogooue on chrom 2RL\n",
      "Computing for 10 samples in cohort CAR_Ombella-MPoko\n",
      "Sample 0 done                    \n",
      "Done for cohort CAR_Ombella-MPoko on chrom 2RL\n",
      "Computing for 45 samples in cohort Cameroon_Adamawa\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Cameroon_Adamawa on chrom 2RL\n",
      "Computing for 30 samples in cohort Ghana_Ashanti-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Done for cohort Ghana_Ashanti-Region on chrom 2RL\n",
      "Computing for 17 samples in cohort Malawi_Southern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Done for cohort Malawi_Southern-Region on chrom 2RL\n",
      "Computing for 89 samples in cohort Mozambique_Maputo\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Sample 50 done                   \n",
      "Sample 60 done                   \n",
      "Sample 70 done                   \n",
      "Sample 80 done                   \n",
      "Done for cohort Mozambique_Maputo on chrom 2RL\n",
      "Computing for 49 samples in cohort Uganda_Eastern-Region\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Uganda_Eastern-Region on chrom 2RL\n",
      "Computing for 35 samples in cohort Benin_Atlantique-Dept\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Benin_Atlantique-Dept on chrom 2RL\n",
      "Computing for 32 samples in cohort DRC_Kinshasa\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort DRC_Kinshasa on chrom 2RL\n",
      "Computing for 41 samples in cohort Nigeria_Ogun-State\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Nigeria_Ogun-State on chrom 2RL\n",
      "Computing for 40 samples in cohort Zambia_Eastern-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Zambia_Eastern-Prov on chrom 2RL\n",
      "Computing for 49 samples in cohort Kenya_Nyanza-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort Kenya_Nyanza-Prov on chrom 2RL\n",
      "Computing for 20 samples in cohort Kenya_Western-Prov\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Done for cohort Kenya_Western-Prov on chrom 2RL\n",
      "Computing for 7 samples in cohort Tanzania_Morogoro-Region\n",
      "Sample 0 done                    \n",
      "Done for cohort Tanzania_Morogoro-Region on chrom 2RL\n",
      "Computing for 43 samples in cohort DRC_Haut-Uele\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Done for cohort DRC_Haut-Uele on chrom 2RL\n",
      "Computing for 40 samples in cohort Mozambique_Cabo-Delgado\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Done for cohort Mozambique_Cabo-Delgado on chrom 2RL\n"
     ]
    }
   ],
   "source": [
    "for chrom in ['X', '3RL', '2RL']:\n",
    "    for pop in sample_idx_dict.keys():\n",
    "        #skip full set for now\n",
    "        if pop != 'hicov':\n",
    "            run_analysis(chrom, pop, sample_idx=sample_idx_dict[pop])\n",
    "            print(f'Done for cohort {pop} on chrom {chrom}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e852ae-335a-4ead-8463-f6dc4df583fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22572d0d-3e3c-49b6-9bdc-c28f77d49c28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2584261b-fe90-4594-b6c1-b1a52950e122",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b86e51-5bd3-42c0-8051-2c9cdda65750",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a21a531d-02d5-46f4-a9df-71ec4ce5dc18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2acf2330-e6ba-4c05-92ae-78c089a846a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "global-global-mgenv-6.0.6",
   "language": "python",
   "name": "conda-env-global-global-mgenv-6.0.6-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
