{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "248b240e-8393-4934-bb7f-4e6fb192bb1a",
   "metadata": {},
   "source": [
    "Get cumulative numbers of variants, targets and genes for hicov geographic cohorts \n",
    "and hicov all "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51459662-9dab-4e66-90bb-56048c62d873",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef9cfdd1-2070-466a-9ef8-e42aee245a1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import allel\n",
    "import collections\n",
    "import numpy as np\n",
    "import malariagen_data\n",
    "import gcsfs\n",
    "import zarr\n",
    "import dask\n",
    "import dask.array as da\n",
    "import cython\n",
    "import functools\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6b0137e-0db0-4898-b4c1-2168a95a9779",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from dask.distributed import Client\n",
    "import dask\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': False}) # Silence large chunk warnings\n",
    "import dask.array as da\n",
    "from dask import delayed, compute\n",
    "from dask_gateway import Gateway\n",
    "import functools\n",
    "import numcodecs\n",
    "from fsspec.implementations.zip import ZipFileSystem\n",
    "from collections.abc import Mapping\n",
    "import gcsfs\n",
    "import numba\n",
    "import psutil\n",
    "from humanize import naturalsize\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f425e08-f531-401a-ae5b-53550cdece7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gcs = gcsfs.GCSFileSystem()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cbfd441-21c4-496b-9795-7c6a50547fb5",
   "metadata": {},
   "source": [
    "## Read ref genome, accessibility and annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b92ce2e-901c-47f0-9aad-35f12fa9dcdd",
   "metadata": {},
   "source": [
    "### Ref genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7da25f9c-fea3-415a-8631-3610622dcf32",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.1.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Load af1 to access reference genome\n",
    "#Sequence will be read for chromosomes separately\n",
    "ag3 = malariagen_data.Ag3(release='3.0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4e5f02-8503-4751-8914-95408a3c29c4",
   "metadata": {},
   "source": [
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "858a6eb6-cf0a-43c7-a7e8-816384803252",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(205568,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22703</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">mRNA</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22533</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002-RA</td><td style=\"text-align: center\">AGAP000002</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">mRNA</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22703</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002-RB</td><td style=\"text-align: center\">AGAP000002</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">205565</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">five_prime_UTR</td><td style=\"text-align: center\">9463141</td><td style=\"text-align: center\">9463142</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R846_3</td><td style=\"text-align: center\">AGAP029992.R846</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">205566</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">9767129</td><td style=\"text-align: center\">9768776</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R845_4</td><td style=\"text-align: center\">AGAP029992.R845</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">205567</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">9767129</td><td style=\"text-align: center\">9768776</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R846_4</td><td style=\"text-align: center\">AGAP029992.R846</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(205568,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AgamP4_X', 'VEuPathDB', 'protein_coding_gene',     582,   22703, -1., '-', -1, 'AGAP000002', '.')\n",
       " ('AgamP4_X', 'VEuPathDB', 'mRNA',     582,   22533, -1., '-', -1, 'AGAP000002-RA', 'AGAP000002')\n",
       " ('AgamP4_X', 'VEuPathDB', 'mRNA',     582,   22703, -1., '-', -1, 'AGAP000002-RB', 'AGAP000002')\n",
       " ...\n",
       " ('AgamP4_3L', 'VEuPathDB', 'five_prime_UTR', 9463141, 9463142, -1., '+', -1, 'utr_AGAP029992.R846_3', 'AGAP029992.R846')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'three_prime_UTR', 9767129, 9768776, -1., '+', -1, 'utr_AGAP029992.R845_4', 'AGAP029992.R845')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'three_prime_UTR', 9767129, 9768776, -1., '+', -1, 'utr_AGAP029992.R846_4', 'AGAP029992.R846')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Uploaded local copy of annotation\n",
    "#Because the release contains a preliminary version\n",
    "gff_fn = 'VectorBase-65_AgambiaePEST.gff'\n",
    "features = allel.FeatureTable.from_gff3(gff_fn, attributes=['ID', 'Parent'])\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a1aca07e-2693-4dbb-9f39-fa5401b6825b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('exon', 75811),\n",
       " ('CDS', 68943),\n",
       " ('five_prime_UTR', 18226),\n",
       " ('mRNA', 15328),\n",
       " ('protein_coding_gene', 13107),\n",
       " ('three_prime_UTR', 12677),\n",
       " ('ncRNA_gene', 729),\n",
       " ('tRNA', 362),\n",
       " ('rRNA', 242),\n",
       " ('pre_miRNA', 77),\n",
       " ('snRNA', 35),\n",
       " ('pseudogene', 9),\n",
       " ('pseudogenic_transcript', 9),\n",
       " ('ncRNA', 4),\n",
       " ('SRP_RNA', 3),\n",
       " ('snoRNA', 2),\n",
       " ('lnc_RNA', 2),\n",
       " ('RNase_P_RNA', 1),\n",
       " ('RNase_MRP_RNA', 1)]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Eyeball the features present\n",
    "collections.Counter(features.type).most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c2c430b5-462c-4992-b8c2-47db9851acca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGAP000002'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# index features by their parent\n",
    "idx_feature_parent = dict(features[['ID', 'Parent']])\n",
    "idx_feature_parent['AGAP000002-RA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4474ab23-2a97-4ec4-83b8-e1d0648b4c31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('AgamP4_2R', 57832),\n",
       " ('AgamP4_2L', 48548),\n",
       " ('AgamP4_3R', 43003),\n",
       " ('AgamP4_3L', 32236),\n",
       " ('AgamP4_X', 19213),\n",
       " ('AgamP4_UNKN', 4593),\n",
       " ('AgamP4_Mt', 124),\n",
       " ('AgamP4_Y_unplaced', 19)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check scaffold names\n",
    "collections.Counter(features.seqid).most_common()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6da04d08-0214-4803-9ae2-6da30b89c992",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(200832,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22703</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">mRNA</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22533</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002-RA</td><td style=\"text-align: center\">AGAP000002</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">mRNA</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22703</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002-RB</td><td style=\"text-align: center\">AGAP000002</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">200829</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">five_prime_UTR</td><td style=\"text-align: center\">9463141</td><td style=\"text-align: center\">9463142</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R846_3</td><td style=\"text-align: center\">AGAP029992.R846</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">200830</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">9767129</td><td style=\"text-align: center\">9768776</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R845_4</td><td style=\"text-align: center\">AGAP029992.R845</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">200831</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">three_prime_UTR</td><td style=\"text-align: center\">9767129</td><td style=\"text-align: center\">9768776</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">utr_AGAP029992.R846_4</td><td style=\"text-align: center\">AGAP029992.R846</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(200832,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AgamP4_X', 'VEuPathDB', 'protein_coding_gene',     582,   22703, -1., '-', -1, 'AGAP000002', '.')\n",
       " ('AgamP4_X', 'VEuPathDB', 'mRNA',     582,   22533, -1., '-', -1, 'AGAP000002-RA', 'AGAP000002')\n",
       " ('AgamP4_X', 'VEuPathDB', 'mRNA',     582,   22703, -1., '-', -1, 'AGAP000002-RB', 'AGAP000002')\n",
       " ...\n",
       " ('AgamP4_3L', 'VEuPathDB', 'five_prime_UTR', 9463141, 9463142, -1., '+', -1, 'utr_AGAP029992.R846_3', 'AGAP029992.R846')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'three_prime_UTR', 9767129, 9768776, -1., '+', -1, 'utr_AGAP029992.R845_4', 'AGAP029992.R845')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'three_prime_UTR', 9767129, 9768776, -1., '+', -1, 'utr_AGAP029992.R846_4', 'AGAP029992.R846')]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to features annotated on the chromosomes\n",
    "loc_features_chroms = (\n",
    "    (features.seqid == 'AgamP4_2R') |\n",
    "    (features.seqid == 'AgamP4_2L') |\n",
    "    (features.seqid == 'AgamP4_3R') |\n",
    "    (features.seqid == 'AgamP4_3L') |\n",
    "    (features.seqid == 'AgamP4_X') \n",
    ")\n",
    "features_chroms = features[loc_features_chroms]\n",
    "features_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0ff8f7c4-3859-4496-8f9e-0ebe7f98a5d0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AgamP4_2L', 'AgamP4_2R', 'AgamP4_3L', 'AgamP4_3R', 'AgamP4_X'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check that it worked\n",
    "np.unique(features_chroms.seqid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fd61ae32-87bb-42f2-85fd-d1083cbaa3b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(67671,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">865</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AGAP000002-PA-CDS5</td><td style=\"text-align: center\">AGAP000002-RA</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">865</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AGAP000002-PB-CDS5</td><td style=\"text-align: center\">AGAP000002-RB</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">950</td><td style=\"text-align: center\">3120</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">1</td><td style=\"text-align: center\">AGAP000002-PA-CDS4</td><td style=\"text-align: center\">AGAP000002-RA</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">67668</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">9766368</td><td style=\"text-align: center\">9766462</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">2</td><td style=\"text-align: center\">AGAP029992.P846-CDS22</td><td style=\"text-align: center\">AGAP029992.R846</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">67669</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">9766577</td><td style=\"text-align: center\">9767128</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AGAP029992.P845-CDS22</td><td style=\"text-align: center\">AGAP029992.R845</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">67670</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">CDS</td><td style=\"text-align: center\">9766577</td><td style=\"text-align: center\">9767128</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">0</td><td style=\"text-align: center\">AGAP029992.P846-CDS23</td><td style=\"text-align: center\">AGAP029992.R846</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(67671,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AgamP4_X', 'VEuPathDB', 'CDS',     582,     865, -1., '-', 2, 'AGAP000002-PA-CDS5', 'AGAP000002-RA')\n",
       " ('AgamP4_X', 'VEuPathDB', 'CDS',     582,     865, -1., '-', 2, 'AGAP000002-PB-CDS5', 'AGAP000002-RB')\n",
       " ('AgamP4_X', 'VEuPathDB', 'CDS',     950,    3120, -1., '-', 1, 'AGAP000002-PA-CDS4', 'AGAP000002-RA')\n",
       " ...\n",
       " ('AgamP4_3L', 'VEuPathDB', 'CDS', 9766368, 9766462, -1., '+', 2, 'AGAP029992.P846-CDS22', 'AGAP029992.R846')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'CDS', 9766577, 9767128, -1., '+', 0, 'AGAP029992.P845-CDS22', 'AGAP029992.R845')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'CDS', 9766577, 9767128, -1., '+', 0, 'AGAP029992.P846-CDS23', 'AGAP029992.R846')]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset to coding sequences on the chromosomes\n",
    "cdss_chroms = features_chroms[features_chroms.type == 'CDS']\n",
    "cdss_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d476e08e-2ea2-4987-82df-2d4d7bc6a5b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAsTable\"><span>&lt;FeatureTable shape=(12614,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '&lt;i8'), ('end', '&lt;i8'), ('score', '&lt;f8'), ('strand', 'O'), ('phase', '&lt;i8'), ('ID', 'O'), ('Parent', 'O')])&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">seqid</th><th style=\"text-align: center\">source</th><th style=\"text-align: center\">type</th><th style=\"text-align: center\">start</th><th style=\"text-align: center\">end</th><th style=\"text-align: center\">score</th><th style=\"text-align: center\">strand</th><th style=\"text-align: center\">phase</th><th style=\"text-align: center\">ID</th><th style=\"text-align: center\">Parent</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">582</td><td style=\"text-align: center\">22703</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000002</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">32382</td><td style=\"text-align: center\">38843</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000005</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">AgamP4_X</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">83817</td><td style=\"text-align: center\">88773</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP000007</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"11\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12611</th><td style=\"text-align: center\">AgamP4_3R</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">35431150</td><td style=\"text-align: center\">35594823</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP029990</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12612</th><td style=\"text-align: center\">AgamP4_2R</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">36940814</td><td style=\"text-align: center\">37029742</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">-</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP029991</td><td style=\"text-align: center\">.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">12613</th><td style=\"text-align: center\">AgamP4_3L</td><td style=\"text-align: center\">VEuPathDB</td><td style=\"text-align: center\">protein_coding_gene</td><td style=\"text-align: center\">9456051</td><td style=\"text-align: center\">9768776</td><td style=\"text-align: center\">-1.0</td><td style=\"text-align: center\">+</td><td style=\"text-align: center\">-1</td><td style=\"text-align: center\">AGAP029992</td><td style=\"text-align: center\">.</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<FeatureTable shape=(12614,) dtype=(numpy.record, [('seqid', 'O'), ('source', 'O'), ('type', 'O'), ('start', '<i8'), ('end', '<i8'), ('score', '<f8'), ('strand', 'O'), ('phase', '<i8'), ('ID', 'O'), ('Parent', 'O')])>\n",
       "[('AgamP4_X', 'VEuPathDB', 'protein_coding_gene',      582,    22703, -1., '-', -1, 'AGAP000002', '.')\n",
       " ('AgamP4_X', 'VEuPathDB', 'protein_coding_gene',    32382,    38843, -1., '-', -1, 'AGAP000005', '.')\n",
       " ('AgamP4_X', 'VEuPathDB', 'protein_coding_gene',    83817,    88773, -1., '-', -1, 'AGAP000007', '.')\n",
       " ...\n",
       " ('AgamP4_3R', 'VEuPathDB', 'protein_coding_gene', 35431150, 35594823, -1., '-', -1, 'AGAP029990', '.')\n",
       " ('AgamP4_2R', 'VEuPathDB', 'protein_coding_gene', 36940814, 37029742, -1., '-', -1, 'AGAP029991', '.')\n",
       " ('AgamP4_3L', 'VEuPathDB', 'protein_coding_gene',  9456051,  9768776, -1., '+', -1, 'AGAP029992', '.')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#subset to protein coding genes on the chromosomes\n",
    "genes_coding_chroms = features_chroms[features_chroms.type =='protein_coding_gene']\n",
    "genes_coding_chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "10bfc7c9-597f-4f22-a454-ecf9b935b6b9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12614"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# double-check that you get the same number of protein coding genes via cds\n",
    "uq_genes_coding_chroms = np.unique([idx_feature_parent[t] for t in cdss_chroms.Parent])\n",
    "assert set(genes_coding_chroms.ID) == set(uq_genes_coding_chroms)\n",
    "len(uq_genes_coding_chroms)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1480579-92b4-4d15-a648-67707ab45748",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2fc8acc3-54e2-427f-93d1-c5d3fe4253ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext cython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b49d53f3-54fd-4281-8e59-0b8ae4b36934",
   "metadata": {},
   "source": [
    "Don't know how to safely use comments in cython - so I'll include them here\n",
    "\n",
    "`%%cython` instructs to compile this cell with cython. the `-a` flag tells it to showcase the compilation.  \n",
    "Next define the nucleotides as integers corresponding to their ascii value.  \n",
    "`@cython.boundscheck(False)` tells cython to not perform a boundscheck -- so we have to guarantee the bounds.  \n",
    "`opt_is_cas9_target()` takes as arguments a 'view' of seq (which we explicitly say cannot be None), a 'view' of is_variant and a boolean specifying whether to check reverse complements. I guess a view is an dynamic array?  \n",
    "Initialise the `out` array, a counter `i` and boleans `fwd`, `rev` and fill `out` with as many zeroes as `seq` is long (but now make it a 1bit integer, whereas it was initialised as an 8bit integer?)  \n",
    "Then release the GIL (it's something of a global lock on python code, and it has to be released to allow for multithreading).  \n",
    "Loop through the positions in seq, taking into account the length of the targetsite. Check that of 21 consecutive positions, none are `N` (except the first of the PAM, which can be anything) and it ends in `GG`. Returns `fwd = True` if this check is satisfied. In that case, that that `is_variant` is zero at all positions except the third last (so if we give a boolean array of where bases are segregating within the population, it gets targets that are non-segregating; but if we give a boolean array of where bases are non-accessbile, it gets targets that are accessible). Repeat in reverse complement if `revcomp = True`.  \n",
    "After finishing the loop, return a view of an array of booleans specifiying for each target starting postition whether its a valid target either forwards or reverse. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4268b4ec-87cf-4c5f-b340-8482d6b69c27",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "\n",
    "import numpy as np\n",
    "cimport numpy as cnp\n",
    "cimport cython\n",
    "\n",
    "cdef:\n",
    "    cnp.uint8_t A = ord('A')\n",
    "    cnp.uint8_t C = ord('C')\n",
    "    cnp.uint8_t G = ord('G')\n",
    "    cnp.uint8_t T = ord('T')\n",
    "    cnp.uint8_t N = ord('N')\n",
    "    \n",
    "@cython.boundscheck(False)\n",
    "def opt_is_cas9_target(cnp.uint8_t[:] seq not None, cnp.uint8_t[:] is_variant, bint revcomp=True):\n",
    "    \"\"\"Locate Cas9 target sequences.\"\"\"\n",
    "    \n",
    "    cdef:\n",
    "        cnp.uint8_t[:] out\n",
    "        Py_ssize_t i\n",
    "        bint fwd, rev\n",
    "        \n",
    "    out = np.zeros(seq.shape[0], dtype='u1')\n",
    "    \n",
    "    with nogil:\n",
    "        for i in range(seq.shape[0] - 20):\n",
    "            # check if forward strand has -NGG\n",
    "            fwd = ((seq[i] != N) and\n",
    "                   (seq[i+1] != N) and\n",
    "                   (seq[i+2] != N) and\n",
    "                   (seq[i+3] != N) and\n",
    "                   (seq[i+4] != N) and\n",
    "                   (seq[i+5] != N) and\n",
    "                   (seq[i+6] != N) and\n",
    "                   (seq[i+7] != N) and\n",
    "                   (seq[i+8] != N) and\n",
    "                   (seq[i+9] != N) and\n",
    "                   (seq[i+10] != N) and\n",
    "                   (seq[i+11] != N) and\n",
    "                   (seq[i+12] != N) and\n",
    "                   (seq[i+13] != N) and\n",
    "                   (seq[i+14] != N) and\n",
    "                   (seq[i+15] != N) and\n",
    "                   (seq[i+16] != N) and\n",
    "                   (seq[i+17] != N) and\n",
    "                   # anything and\n",
    "                   (seq[i+19] == G) and \n",
    "                   (seq[i+20] == G))\n",
    "            if fwd and is_variant is not None:\n",
    "                fwd = (fwd and \n",
    "                       (is_variant[i] == 0) and\n",
    "                       (is_variant[i+1] == 0) and\n",
    "                       (is_variant[i+2] == 0) and\n",
    "                       (is_variant[i+3] == 0) and\n",
    "                       (is_variant[i+4] == 0) and\n",
    "                       (is_variant[i+5] == 0) and\n",
    "                       (is_variant[i+6] == 0) and\n",
    "                       (is_variant[i+7] == 0) and\n",
    "                       (is_variant[i+8] == 0) and\n",
    "                       (is_variant[i+9] == 0) and\n",
    "                       (is_variant[i+10] == 0) and\n",
    "                       (is_variant[i+11] == 0) and\n",
    "                       (is_variant[i+12] == 0) and\n",
    "                       (is_variant[i+13] == 0) and\n",
    "                       (is_variant[i+14] == 0) and\n",
    "                       (is_variant[i+15] == 0) and\n",
    "                       (is_variant[i+16] == 0) and\n",
    "                       (is_variant[i+17] == 0) and\n",
    "                       # anything and\n",
    "                       (is_variant[i+19] == 0) and\n",
    "                       (is_variant[i+20] == 0))\n",
    "            if revcomp:\n",
    "                # check if reverse strand has -NGG\n",
    "                rev = ((seq[i] == C) and\n",
    "                       (seq[i+1] == C) and\n",
    "                       # anything and\n",
    "                       (seq[i+3] != N) and\n",
    "                       (seq[i+4] != N) and\n",
    "                       (seq[i+5] != N) and\n",
    "                       (seq[i+6] != N) and\n",
    "                       (seq[i+7] != N) and\n",
    "                       (seq[i+8] != N) and\n",
    "                       (seq[i+9] != N) and\n",
    "                       (seq[i+10] != N) and\n",
    "                       (seq[i+11] != N) and\n",
    "                       (seq[i+12] != N) and\n",
    "                       (seq[i+13] != N) and\n",
    "                       (seq[i+14] != N) and\n",
    "                       (seq[i+15] != N) and\n",
    "                       (seq[i+16] != N) and\n",
    "                       (seq[i+17] != N) and\n",
    "                       (seq[i+18] != N) and\n",
    "                       (seq[i+19] != N) and \n",
    "                       (seq[i+20] != N))\n",
    "                if rev and is_variant is not None:\n",
    "                    rev = (rev and \n",
    "                           (is_variant[i] == 0) and\n",
    "                           (is_variant[i+1] == 0) and\n",
    "                           # anything and\n",
    "                           (is_variant[i+3] == 0) and\n",
    "                           (is_variant[i+4] == 0) and\n",
    "                           (is_variant[i+5] == 0) and\n",
    "                           (is_variant[i+6] == 0) and\n",
    "                           (is_variant[i+7] == 0) and\n",
    "                           (is_variant[i+8] == 0) and\n",
    "                           (is_variant[i+9] == 0) and\n",
    "                           (is_variant[i+10] == 0) and\n",
    "                           (is_variant[i+11] == 0) and\n",
    "                           (is_variant[i+12] == 0) and\n",
    "                           (is_variant[i+13] == 0) and\n",
    "                           (is_variant[i+14] == 0) and\n",
    "                           (is_variant[i+15] == 0) and\n",
    "                           (is_variant[i+16] == 0) and\n",
    "                           (is_variant[i+17] == 0) and\n",
    "                           (is_variant[i+18] == 0) and\n",
    "                           (is_variant[i+19] == 0) and\n",
    "                           (is_variant[i+20] == 0))\n",
    "            else:\n",
    "                rev = False\n",
    "            out[i] = fwd or rev\n",
    "    \n",
    "    return np.asarray(out).view(bool)\n",
    "\n",
    "@cython.boundscheck(False)\n",
    "def opt_all_subsequent(cython.integral[:] t, Py_ssize_t n):\n",
    "    \"\"\"Locate contiguous regions with the same non-zero value.\"\"\"\n",
    "    \n",
    "    cdef:\n",
    "        cython.integral[:] out\n",
    "        Py_ssize_t i, j\n",
    "        cython.integral x\n",
    "        bint subs\n",
    "        \n",
    "    out = np.zeros_like(t)\n",
    "    \n",
    "    with nogil:\n",
    "        for i in range(t.shape[0] - n):\n",
    "            x = t[i]\n",
    "            if x > 0:\n",
    "                subs = True\n",
    "                for j in range(i+1, i+n):\n",
    "                    subs = subs and t[j] == x\n",
    "                if subs:\n",
    "                    out[i] = x\n",
    "                \n",
    "    return np.asarray(out)\n",
    "  \n",
    "@cython.boundscheck(False)\n",
    "def opt_zero_subsequent(cython.integral[:] t, Py_ssize_t n):\n",
    "    \"\"\"Locate non-overlapping targets.\"\"\"\n",
    "\n",
    "    cdef:\n",
    "        cython.integral[:] out\n",
    "        Py_ssize_t i, j\n",
    "        cython.integral x\n",
    "        \n",
    "    out = np.zeros_like(t)\n",
    "    i = 0\n",
    "\n",
    "    with nogil:\n",
    "        while i < t.shape[0]:\n",
    "            x = t[i]\n",
    "            if x > 0:\n",
    "                out[i] = x\n",
    "                i += n\n",
    "            else:\n",
    "                i += 1\n",
    "\n",
    "    return np.asarray(out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d10e753-c0f1-45c8-8814-51ba3a5b965e",
   "metadata": {},
   "source": [
    "`opt_all_subsequent()` scans an array of integer values and reports for each entry, whether the `n` subsequent entries have the same non-zero value (used to check e.g. whether all entries are in the same exon). If the requirement is met, it replaces the corresponding entry in `out` by its value in `t`. Returns numpy array `out`, with zeroes where the condition is not met.  \n",
    "`opt_zero_subsequent()` scans an array of integer values and returns an array of integers where non-zero values are spaces apart by at least `n`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "866cea0d-cb76-4f71-a114-be949460f777",
   "metadata": {},
   "source": [
    "### Python functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9f66d875-86ba-45e2-abb0-244f0714227d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_reference_sequence(chrom):\n",
    "    \"\"\"Load the reference sequence.\"\"\"\n",
    "    seq = ag3.genome_sequence(chrom).compute()\n",
    "    seq = np.char.upper(seq)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bb9b2a46-79f3-4fd1-ad86-07ef1858eaec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# can't cache, but it's pretty quick\n",
    "def get_is_cas9_target(chrom, revcomp=True, is_variant=None):\n",
    "    \"\"\"Find Cas9 target sequences.\"\"\"\n",
    "    seq = get_reference_sequence(chrom).view('u1')\n",
    "    if is_variant is not None:\n",
    "        is_variant = is_variant.view('u1')\n",
    "    out = opt_is_cas9_target(seq, revcomp=revcomp, is_variant=is_variant)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0461b858-9416-45cb-9567-061224b4169d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_cds_idx(chrom):\n",
    "    \"\"\"Locate CDS features on the reference sequence.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=int)\n",
    "    chrom = f'AgamP4_{chrom}'\n",
    "    for i, (seqid, start, end) in enumerate(cdss_chroms[['seqid', 'start', 'end']]):\n",
    "        if seqid == chrom:\n",
    "            out[start-1:end] = i + 1  # start indexing from 1\n",
    "    return out\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f81dcc3-8ca7-4ae6-8b40-4e0c186fdf9d",
   "metadata": {},
   "source": [
    "This gives each CDS a unique index -- however, if there are ovelapping CDS, the last listed one counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e4e1cf85-fe3a-4349-950b-0a97c9d63d85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_gene_idx(chrom):\n",
    "    \"\"\"Locate protein-coding genes on the reference sequence.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=int)\n",
    "    chrom = f'AgamP4_{chrom}'\n",
    "    for i, (seqid, start, end) in enumerate(genes_coding_chroms[['seqid', 'start', 'end']]):\n",
    "        if seqid == chrom:\n",
    "            out[start-1:end] = i + 1  # start indexing from 1\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aef75df5-45e3-4890-b7de-f6de70d5b254",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def get_target_in_cds(chrom):\n",
    "    \"\"\"Locate targets within a single CDS.\"\"\"\n",
    "    t = get_cds_idx(chrom)\n",
    "    out = opt_all_subsequent(t, 21) > 0\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "242ba5cb-ca03-4041-9e12-540b2f69736d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#@functools.lru_cache(maxsize=None)\n",
    "def get_is_variant_individual(chrom, sample_idx):\n",
    "    \"\"\"Locate variant sites for a given sample.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=bool)\n",
    "    pos = ag3.snp_calls(chrom, sample_sets='3.0').variant_position.values\n",
    "    sample_idx_list = [int(sample_idx)]\n",
    "    g = ag3.snp_calls(chrom, sample_sets = ['AG1000G-AO', 'AG1000G-GW', \n",
    "                                            'AG1000G-BF-A', 'AG1000G-GN-A', \n",
    "               'AG1000G-CM-A', 'AG1000G-GA-A', 'AG1000G-UG', 'AG1000G-KE'], \n",
    "                      sample_indices = sample_idx_list).call_genotype\n",
    "    loc = np.any(g > 0, axis=(1,2))\n",
    "    loc = loc.compute().values\n",
    "    idx = pos[loc] - 1\n",
    "    out[idx] = True\n",
    "\n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd59838f-9636-4297-a709-9da7ca97c213",
   "metadata": {},
   "source": [
    "# Set up cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "45a72b46-3efa-4451-b9fd-3ab3ed92c777",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gateway = Gateway()\n",
    "for cl in gateway.list_clusters():\n",
    "    gateway.connect(cl.name).shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ba6228b3-54a9-4fb4-8e46-fe062b66433f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "023f0f965e9e452cb6a0291491b75521",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HTML(value='<h2>GatewayCluster</h2>'), HBox(children=(HTML(value='\\n<div>\\n<style scoped>\\n    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gateway = Gateway()\n",
    "conda_prefix = os.environ[\"CONDA_PREFIX\"]\n",
    "current_environment = 'global/'+conda_prefix.split('/')[5]\n",
    "cluster = gateway.new_cluster(\n",
    "    profile='standard', \n",
    "    conda_environment = current_environment,\n",
    ")\n",
    "cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f7b69903-d74a-48fc-91b5-de5893ef8392",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "client=cluster.get_client()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "268019bf-188c-4eb0-a905-e09f58e807ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cluster.scale(60)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fbd79e1-81b7-4027-b9c1-946c6738d500",
   "metadata": {},
   "source": [
    "# Run analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fde3d6e4-7ac3-4353-9599-75a43c1de752",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def identify_targets_population_cumulative(chrom, sample_idx):\n",
    "\n",
    "    #Find targets in cds (not using site filter)\n",
    "    base_loc = get_target_in_cds(chrom) \n",
    "    loc = base_loc & get_is_cas9_target(chrom, revcomp=True)\n",
    "    \n",
    "    #index genes\n",
    "    gene_idx = get_gene_idx(chrom)\n",
    "    \n",
    "    # setup outputs\n",
    "    out_sites = []\n",
    "    out_targets = []\n",
    "    out_genes = []\n",
    "    \n",
    "    # setup variation\n",
    "    is_variant = np.zeros(gene_idx.shape[0], dtype=bool)\n",
    "    \n",
    "    #initial numbers\n",
    "    out_sites.append(0)\n",
    "    \n",
    "    n_targets = np.count_nonzero(loc)\n",
    "    out_targets.append(n_targets)\n",
    "    \n",
    "    n_genes = len(np.unique(gene_idx[loc]))\n",
    "    out_genes.append(n_genes)\n",
    "    \n",
    "    #randomise order of samples\n",
    "    rng = np.random.default_rng()\n",
    "    sample_idx = rng.choice(sample_idx, size=len(sample_idx),\n",
    "                              replace=False)\n",
    "    \n",
    "    for i, sidx in enumerate(sample_idx):\n",
    "        \n",
    "        #accumulate variation\n",
    "        is_variant |= get_is_variant_individual(chrom, sidx)\n",
    "        \n",
    "        # locate sites\n",
    "        n_sites = np.count_nonzero(is_variant)\n",
    "        out_sites.append(n_sites)\n",
    "        \n",
    "        # locate targets and genes\n",
    "        loc = base_loc & get_is_cas9_target(chrom, revcomp=True, is_variant=is_variant)\n",
    "        n_targets = np.count_nonzero(loc)\n",
    "        out_targets.append(n_targets)\n",
    "\n",
    "        n_genes = len(np.unique(gene_idx[loc]))\n",
    "        out_genes.append(n_genes)\n",
    "        if i%10 == 0:\n",
    "            print(f'Sample {i} done')\n",
    "\n",
    "\n",
    "    return np.array([out_sites, out_targets, out_genes]).T\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "44475a5b-b2a3-44b1-ab88-74923b9d69df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_analysis(chrom, pop, sample_idx):\n",
    "    \n",
    "    outdir = f'cumulative_results_gam/{pop}/'\n",
    "    if os.path.exists(f'{outdir}/target_info_{chrom}.npy'): \n",
    "        print(f'{outdir}/target_info_{chrom}.npy already exists, skipping computation')\n",
    "    else:\n",
    "        if not os.path.isdir(outdir):\n",
    "            !mkdir {outdir}\n",
    "        \n",
    "        print(f'Computing for {len(sample_idx)} samples in cohort {pop}')\n",
    "        target_info = identify_targets_population_cumulative(chrom, sample_idx)\n",
    "        np.save(f'{outdir}/target_info_{chrom}.npy', target_info)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "06bdd6ed-69a7-438a-89f9-c7c26f801417",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#The sample sets that contain Ag1 samples\n",
    "sample_sets = ['AG1000G-AO', 'AG1000G-GW', 'AG1000G-BF-A', 'AG1000G-GN-A', \n",
    "               'AG1000G-CM-A', 'AG1000G-GA-A', 'AG1000G-UG', 'AG1000G-KE']\n",
    "#The sample ids of Ag1 samples\n",
    "sample_ids = pd.read_csv('ag1_sample_ids.csv')\n",
    "sample_idsx = sample_ids.sample_id + 'x'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2d58e132-57cc-4d2e-b7b7-744589c79136",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    }
   ],
   "source": [
    "#Get metadata subsetted to Ag1 samples\n",
    "meta_p1 = ag3.sample_metadata(sample_sets = sample_sets)\n",
    "meta_p1 = meta_p1.loc[meta_p1.sample_id.isin(sample_ids.sample_id)|meta_p1.sample_id.isin(sample_idsx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2e2e0370-71ce-4f66-b953-480772d54998",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(762, 32)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_p1.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa6daa25-b4fa-460d-9060-fcf67140c334",
   "metadata": {},
   "source": [
    "Three samples missing: 1 from Angola and 2 from Kenya. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4605ed9b-e278-4da3-b0b4-870d4c9a3307",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Overwrite sample_ids to use it for sample index selection\n",
    "sample_ids = meta_p1.sample_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "603c50ad-67c4-41dc-99fb-ebff25924624",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sample_idx_dict = dict({'phase1': tuple(meta_p1.index.values)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a19d58df-7b54-480c-aa6d-a9ee39cffed4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'phase1': (0,\n",
       "  1,\n",
       "  2,\n",
       "  3,\n",
       "  4,\n",
       "  5,\n",
       "  6,\n",
       "  8,\n",
       "  10,\n",
       "  11,\n",
       "  12,\n",
       "  13,\n",
       "  15,\n",
       "  17,\n",
       "  18,\n",
       "  19,\n",
       "  20,\n",
       "  21,\n",
       "  22,\n",
       "  23,\n",
       "  24,\n",
       "  25,\n",
       "  26,\n",
       "  27,\n",
       "  28,\n",
       "  29,\n",
       "  30,\n",
       "  31,\n",
       "  33,\n",
       "  34,\n",
       "  35,\n",
       "  37,\n",
       "  38,\n",
       "  39,\n",
       "  40,\n",
       "  41,\n",
       "  42,\n",
       "  43,\n",
       "  44,\n",
       "  46,\n",
       "  47,\n",
       "  49,\n",
       "  51,\n",
       "  53,\n",
       "  54,\n",
       "  55,\n",
       "  56,\n",
       "  57,\n",
       "  58,\n",
       "  59,\n",
       "  69,\n",
       "  70,\n",
       "  71,\n",
       "  72,\n",
       "  74,\n",
       "  75,\n",
       "  77,\n",
       "  79,\n",
       "  80,\n",
       "  81,\n",
       "  83,\n",
       "  84,\n",
       "  85,\n",
       "  86,\n",
       "  87,\n",
       "  88,\n",
       "  89,\n",
       "  90,\n",
       "  92,\n",
       "  93,\n",
       "  94,\n",
       "  95,\n",
       "  96,\n",
       "  97,\n",
       "  98,\n",
       "  99,\n",
       "  100,\n",
       "  101,\n",
       "  102,\n",
       "  103,\n",
       "  104,\n",
       "  107,\n",
       "  109,\n",
       "  111,\n",
       "  112,\n",
       "  113,\n",
       "  114,\n",
       "  115,\n",
       "  116,\n",
       "  117,\n",
       "  118,\n",
       "  121,\n",
       "  122,\n",
       "  123,\n",
       "  124,\n",
       "  125,\n",
       "  126,\n",
       "  127,\n",
       "  128,\n",
       "  129,\n",
       "  130,\n",
       "  132,\n",
       "  133,\n",
       "  134,\n",
       "  135,\n",
       "  137,\n",
       "  138,\n",
       "  139,\n",
       "  140,\n",
       "  141,\n",
       "  142,\n",
       "  143,\n",
       "  144,\n",
       "  146,\n",
       "  148,\n",
       "  150,\n",
       "  151,\n",
       "  152,\n",
       "  153,\n",
       "  154,\n",
       "  155,\n",
       "  156,\n",
       "  157,\n",
       "  158,\n",
       "  159,\n",
       "  160,\n",
       "  162,\n",
       "  163,\n",
       "  164,\n",
       "  165,\n",
       "  166,\n",
       "  167,\n",
       "  168,\n",
       "  169,\n",
       "  170,\n",
       "  171,\n",
       "  172,\n",
       "  173,\n",
       "  177,\n",
       "  178,\n",
       "  179,\n",
       "  181,\n",
       "  182,\n",
       "  183,\n",
       "  184,\n",
       "  185,\n",
       "  186,\n",
       "  187,\n",
       "  188,\n",
       "  189,\n",
       "  190,\n",
       "  191,\n",
       "  192,\n",
       "  193,\n",
       "  197,\n",
       "  199,\n",
       "  200,\n",
       "  201,\n",
       "  202,\n",
       "  203,\n",
       "  205,\n",
       "  206,\n",
       "  207,\n",
       "  208,\n",
       "  210,\n",
       "  212,\n",
       "  213,\n",
       "  214,\n",
       "  215,\n",
       "  216,\n",
       "  217,\n",
       "  218,\n",
       "  219,\n",
       "  220,\n",
       "  221,\n",
       "  222,\n",
       "  223,\n",
       "  224,\n",
       "  227,\n",
       "  228,\n",
       "  229,\n",
       "  230,\n",
       "  231,\n",
       "  233,\n",
       "  234,\n",
       "  235,\n",
       "  237,\n",
       "  238,\n",
       "  239,\n",
       "  240,\n",
       "  241,\n",
       "  242,\n",
       "  243,\n",
       "  244,\n",
       "  245,\n",
       "  247,\n",
       "  248,\n",
       "  249,\n",
       "  250,\n",
       "  251,\n",
       "  253,\n",
       "  254,\n",
       "  255,\n",
       "  256,\n",
       "  257,\n",
       "  258,\n",
       "  259,\n",
       "  260,\n",
       "  261,\n",
       "  262,\n",
       "  263,\n",
       "  264,\n",
       "  265,\n",
       "  266,\n",
       "  267,\n",
       "  269,\n",
       "  271,\n",
       "  272,\n",
       "  273,\n",
       "  274,\n",
       "  275,\n",
       "  277,\n",
       "  278,\n",
       "  279,\n",
       "  280,\n",
       "  281,\n",
       "  282,\n",
       "  283,\n",
       "  284,\n",
       "  285,\n",
       "  286,\n",
       "  287,\n",
       "  288,\n",
       "  289,\n",
       "  290,\n",
       "  291,\n",
       "  292,\n",
       "  293,\n",
       "  294,\n",
       "  295,\n",
       "  296,\n",
       "  297,\n",
       "  298,\n",
       "  300,\n",
       "  301,\n",
       "  302,\n",
       "  303,\n",
       "  304,\n",
       "  305,\n",
       "  306,\n",
       "  307,\n",
       "  308,\n",
       "  309,\n",
       "  310,\n",
       "  311,\n",
       "  312,\n",
       "  313,\n",
       "  314,\n",
       "  317,\n",
       "  318,\n",
       "  319,\n",
       "  320,\n",
       "  321,\n",
       "  322,\n",
       "  323,\n",
       "  324,\n",
       "  325,\n",
       "  326,\n",
       "  327,\n",
       "  328,\n",
       "  329,\n",
       "  330,\n",
       "  331,\n",
       "  333,\n",
       "  334,\n",
       "  335,\n",
       "  336,\n",
       "  337,\n",
       "  338,\n",
       "  339,\n",
       "  340,\n",
       "  341,\n",
       "  342,\n",
       "  343,\n",
       "  344,\n",
       "  345,\n",
       "  346,\n",
       "  347,\n",
       "  348,\n",
       "  349,\n",
       "  350,\n",
       "  351,\n",
       "  352,\n",
       "  353,\n",
       "  354,\n",
       "  355,\n",
       "  356,\n",
       "  357,\n",
       "  358,\n",
       "  359,\n",
       "  360,\n",
       "  361,\n",
       "  362,\n",
       "  363,\n",
       "  364,\n",
       "  365,\n",
       "  366,\n",
       "  367,\n",
       "  368,\n",
       "  369,\n",
       "  372,\n",
       "  373,\n",
       "  374,\n",
       "  375,\n",
       "  376,\n",
       "  377,\n",
       "  378,\n",
       "  379,\n",
       "  380,\n",
       "  381,\n",
       "  382,\n",
       "  383,\n",
       "  384,\n",
       "  386,\n",
       "  387,\n",
       "  388,\n",
       "  389,\n",
       "  390,\n",
       "  391,\n",
       "  392,\n",
       "  393,\n",
       "  395,\n",
       "  398,\n",
       "  400,\n",
       "  401,\n",
       "  402,\n",
       "  403,\n",
       "  404,\n",
       "  405,\n",
       "  406,\n",
       "  407,\n",
       "  408,\n",
       "  409,\n",
       "  410,\n",
       "  411,\n",
       "  412,\n",
       "  413,\n",
       "  414,\n",
       "  415,\n",
       "  416,\n",
       "  417,\n",
       "  418,\n",
       "  419,\n",
       "  420,\n",
       "  421,\n",
       "  422,\n",
       "  423,\n",
       "  424,\n",
       "  425,\n",
       "  426,\n",
       "  427,\n",
       "  428,\n",
       "  429,\n",
       "  430,\n",
       "  431,\n",
       "  432,\n",
       "  433,\n",
       "  434,\n",
       "  435,\n",
       "  436,\n",
       "  437,\n",
       "  438,\n",
       "  439,\n",
       "  440,\n",
       "  441,\n",
       "  442,\n",
       "  443,\n",
       "  445,\n",
       "  446,\n",
       "  447,\n",
       "  448,\n",
       "  449,\n",
       "  450,\n",
       "  451,\n",
       "  452,\n",
       "  453,\n",
       "  454,\n",
       "  455,\n",
       "  456,\n",
       "  457,\n",
       "  458,\n",
       "  459,\n",
       "  460,\n",
       "  461,\n",
       "  462,\n",
       "  463,\n",
       "  464,\n",
       "  465,\n",
       "  466,\n",
       "  467,\n",
       "  468,\n",
       "  469,\n",
       "  470,\n",
       "  471,\n",
       "  472,\n",
       "  473,\n",
       "  474,\n",
       "  475,\n",
       "  476,\n",
       "  477,\n",
       "  478,\n",
       "  480,\n",
       "  481,\n",
       "  482,\n",
       "  483,\n",
       "  484,\n",
       "  485,\n",
       "  486,\n",
       "  487,\n",
       "  488,\n",
       "  489,\n",
       "  490,\n",
       "  491,\n",
       "  492,\n",
       "  493,\n",
       "  494,\n",
       "  495,\n",
       "  496,\n",
       "  497,\n",
       "  498,\n",
       "  499,\n",
       "  500,\n",
       "  501,\n",
       "  502,\n",
       "  503,\n",
       "  504,\n",
       "  505,\n",
       "  506,\n",
       "  507,\n",
       "  508,\n",
       "  509,\n",
       "  511,\n",
       "  512,\n",
       "  513,\n",
       "  514,\n",
       "  515,\n",
       "  517,\n",
       "  518,\n",
       "  519,\n",
       "  520,\n",
       "  523,\n",
       "  525,\n",
       "  526,\n",
       "  527,\n",
       "  528,\n",
       "  529,\n",
       "  530,\n",
       "  531,\n",
       "  533,\n",
       "  534,\n",
       "  535,\n",
       "  537,\n",
       "  538,\n",
       "  539,\n",
       "  540,\n",
       "  541,\n",
       "  542,\n",
       "  543,\n",
       "  544,\n",
       "  546,\n",
       "  547,\n",
       "  548,\n",
       "  550,\n",
       "  551,\n",
       "  552,\n",
       "  553,\n",
       "  555,\n",
       "  556,\n",
       "  557,\n",
       "  558,\n",
       "  560,\n",
       "  561,\n",
       "  562,\n",
       "  564,\n",
       "  565,\n",
       "  566,\n",
       "  567,\n",
       "  568,\n",
       "  570,\n",
       "  571,\n",
       "  572,\n",
       "  573,\n",
       "  574,\n",
       "  575,\n",
       "  576,\n",
       "  577,\n",
       "  578,\n",
       "  579,\n",
       "  580,\n",
       "  581,\n",
       "  582,\n",
       "  583,\n",
       "  584,\n",
       "  585,\n",
       "  586,\n",
       "  588,\n",
       "  589,\n",
       "  591,\n",
       "  593,\n",
       "  594,\n",
       "  595,\n",
       "  596,\n",
       "  597,\n",
       "  598,\n",
       "  599,\n",
       "  600,\n",
       "  603,\n",
       "  604,\n",
       "  605,\n",
       "  607,\n",
       "  609,\n",
       "  611,\n",
       "  612,\n",
       "  613,\n",
       "  614,\n",
       "  615,\n",
       "  617,\n",
       "  618,\n",
       "  621,\n",
       "  622,\n",
       "  623,\n",
       "  624,\n",
       "  625,\n",
       "  626,\n",
       "  627,\n",
       "  628,\n",
       "  629,\n",
       "  630,\n",
       "  631,\n",
       "  632,\n",
       "  635,\n",
       "  638,\n",
       "  640,\n",
       "  641,\n",
       "  642,\n",
       "  643,\n",
       "  644,\n",
       "  645,\n",
       "  646,\n",
       "  647,\n",
       "  648,\n",
       "  649,\n",
       "  650,\n",
       "  651,\n",
       "  652,\n",
       "  653,\n",
       "  654,\n",
       "  655,\n",
       "  656,\n",
       "  659,\n",
       "  660,\n",
       "  661,\n",
       "  662,\n",
       "  663,\n",
       "  664,\n",
       "  672,\n",
       "  673,\n",
       "  674,\n",
       "  676,\n",
       "  677,\n",
       "  678,\n",
       "  679,\n",
       "  680,\n",
       "  681,\n",
       "  682,\n",
       "  683,\n",
       "  685,\n",
       "  686,\n",
       "  687,\n",
       "  688,\n",
       "  689,\n",
       "  690,\n",
       "  691,\n",
       "  692,\n",
       "  694,\n",
       "  695,\n",
       "  696,\n",
       "  697,\n",
       "  698,\n",
       "  699,\n",
       "  700,\n",
       "  701,\n",
       "  702,\n",
       "  703,\n",
       "  704,\n",
       "  706,\n",
       "  707,\n",
       "  708,\n",
       "  709,\n",
       "  710,\n",
       "  711,\n",
       "  712,\n",
       "  713,\n",
       "  715,\n",
       "  716,\n",
       "  717,\n",
       "  718,\n",
       "  719,\n",
       "  720,\n",
       "  721,\n",
       "  722,\n",
       "  723,\n",
       "  724,\n",
       "  725,\n",
       "  726,\n",
       "  729,\n",
       "  730,\n",
       "  795,\n",
       "  796,\n",
       "  797,\n",
       "  798,\n",
       "  799,\n",
       "  801,\n",
       "  802,\n",
       "  803,\n",
       "  804,\n",
       "  805,\n",
       "  806,\n",
       "  807,\n",
       "  808,\n",
       "  810,\n",
       "  811,\n",
       "  812,\n",
       "  813,\n",
       "  814,\n",
       "  815,\n",
       "  816,\n",
       "  817,\n",
       "  818,\n",
       "  819,\n",
       "  820,\n",
       "  821,\n",
       "  822,\n",
       "  823,\n",
       "  824,\n",
       "  825,\n",
       "  826,\n",
       "  828,\n",
       "  829,\n",
       "  830,\n",
       "  832,\n",
       "  833,\n",
       "  837,\n",
       "  838,\n",
       "  839,\n",
       "  840,\n",
       "  841,\n",
       "  842,\n",
       "  843,\n",
       "  948,\n",
       "  949,\n",
       "  950,\n",
       "  951,\n",
       "  952,\n",
       "  953,\n",
       "  954,\n",
       "  955,\n",
       "  956,\n",
       "  957,\n",
       "  958,\n",
       "  959,\n",
       "  960,\n",
       "  961,\n",
       "  962,\n",
       "  964,\n",
       "  965,\n",
       "  966,\n",
       "  967,\n",
       "  968,\n",
       "  969,\n",
       "  970,\n",
       "  971,\n",
       "  972,\n",
       "  973,\n",
       "  974,\n",
       "  975,\n",
       "  976,\n",
       "  977,\n",
       "  978,\n",
       "  979,\n",
       "  980,\n",
       "  981,\n",
       "  982,\n",
       "  983,\n",
       "  984,\n",
       "  985,\n",
       "  986,\n",
       "  987,\n",
       "  988,\n",
       "  989,\n",
       "  990,\n",
       "  991,\n",
       "  993,\n",
       "  994,\n",
       "  995,\n",
       "  996,\n",
       "  997,\n",
       "  998,\n",
       "  1000,\n",
       "  1001,\n",
       "  1002,\n",
       "  1003,\n",
       "  1005,\n",
       "  1006,\n",
       "  1007,\n",
       "  1008,\n",
       "  1009,\n",
       "  1010,\n",
       "  1011,\n",
       "  1012,\n",
       "  1014,\n",
       "  1016,\n",
       "  1017,\n",
       "  1018,\n",
       "  1019,\n",
       "  1020,\n",
       "  1021,\n",
       "  1022,\n",
       "  1023,\n",
       "  1024,\n",
       "  1025,\n",
       "  1026,\n",
       "  1027,\n",
       "  1028,\n",
       "  1029,\n",
       "  1030,\n",
       "  1031,\n",
       "  1032,\n",
       "  1034,\n",
       "  1035,\n",
       "  1036,\n",
       "  1037,\n",
       "  1038,\n",
       "  1039,\n",
       "  1040,\n",
       "  1042,\n",
       "  1043,\n",
       "  1044,\n",
       "  1045,\n",
       "  1046,\n",
       "  1047,\n",
       "  1048,\n",
       "  1049,\n",
       "  1050,\n",
       "  1051,\n",
       "  1052,\n",
       "  1053,\n",
       "  1055,\n",
       "  1056,\n",
       "  1057,\n",
       "  1058,\n",
       "  1059)}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_idx_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1adb0896-359f-44a5-8705-6f5353d74ac2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cumulative_results_gam/phase1//target_info_X.npy already exists, skipping computation\n",
      "Done for cohort phase1 on chrom X\n",
      "cumulative_results_gam/phase1//target_info_3R.npy already exists, skipping computation\n",
      "Done for cohort phase1 on chrom 3R\n",
      "cumulative_results_gam/phase1//target_info_3L.npy already exists, skipping computation\n",
      "Done for cohort phase1 on chrom 3L\n",
      "Computing for 762 samples in cohort phase1\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Sample 50 done                   \n",
      "Sample 60 done                   \n",
      "Sample 70 done                   \n",
      "Sample 80 done                   \n",
      "                                 \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-13 11:32:35,006 - distributed.client - WARNING - Couldn't gather 3 keys, rescheduling (('array-18cd2ba69f7b41940eed10e41977d3e6', 52), ('array-18cd2ba69f7b41940eed10e41977d3e6', 81), ('array-18cd2ba69f7b41940eed10e41977d3e6', 99))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample 90 done                   \n",
      "Sample 100 done                  \n",
      "Sample 110 done                  \n",
      "Sample 120 done                  \n",
      "Sample 130 done                  \n",
      "Sample 140 done                  \n",
      "Sample 150 done                  \n",
      "Sample 160 done                  \n",
      "Sample 170 done                  \n",
      "Sample 180 done                  \n",
      "Sample 190 done                  \n",
      "Sample 200 done                  \n",
      "Sample 210 done                  \n",
      "Sample 220 done                  \n",
      "Sample 230 done                  \n",
      "Sample 240 done                  \n",
      "Sample 250 done                  \n",
      "Sample 260 done                  \n",
      "Sample 270 done                  \n",
      "Sample 280 done                  \n",
      "Sample 290 done                  \n",
      "Sample 300 done                  \n",
      "Sample 310 done                  \n",
      "Sample 320 done                  \n",
      "Sample 330 done                  \n",
      "Sample 340 done                  \n",
      "Sample 350 done                  \n",
      "Sample 360 done                  \n",
      "Sample 370 done                  \n",
      "Sample 380 done                  \n",
      "Sample 390 done                  \n",
      "Sample 400 done                  \n",
      "Sample 410 done                  \n",
      "Sample 420 done                  \n",
      "Sample 430 done                  \n",
      "Sample 440 done                  \n",
      "Sample 450 done                  \n",
      "Sample 460 done                  \n",
      "Sample 470 done                  \n",
      "Sample 480 done                  \n",
      "Sample 490 done                  \n",
      "Sample 500 done                  \n",
      "Sample 510 done                  \n",
      "Sample 520 done                  \n",
      "Sample 530 done                  \n",
      "Sample 540 done                  \n",
      "Sample 550 done                  \n",
      "Sample 560 done                  \n",
      "Sample 570 done                  \n",
      "Sample 580 done                  \n",
      "Sample 590 done                  \n",
      "Sample 600 done                  \n",
      "Sample 610 done                  \n",
      "Sample 620 done                  \n",
      "Sample 630 done                  \n",
      "Sample 640 done                  \n",
      "Sample 650 done                  \n",
      "Sample 660 done                  \n",
      "Sample 670 done                  \n",
      "Sample 680 done                  \n",
      "Sample 690 done                  \n",
      "Sample 700 done                  \n",
      "Sample 710 done                  \n",
      "Sample 720 done                  \n",
      "Sample 730 done                  \n",
      "Sample 740 done                  \n",
      "Sample 750 done                  \n",
      "Sample 760 done                  \n",
      "Done for cohort phase1 on chrom 2R\n",
      "Computing for 762 samples in cohort phase1\n",
      "Sample 0 done                    \n",
      "Sample 10 done                   \n",
      "Sample 20 done                   \n",
      "Sample 30 done                   \n",
      "Sample 40 done                   \n",
      "Sample 50 done                   \n",
      "Sample 60 done                   \n",
      "Sample 70 done                   \n",
      "Sample 80 done                   \n",
      "Sample 90 done                   \n",
      "Sample 100 done                  \n",
      "Sample 110 done                  \n",
      "Sample 120 done                  \n",
      "Sample 130 done                  \n",
      "Sample 140 done                  \n",
      "Sample 150 done                  \n",
      "Sample 160 done                  \n",
      "Sample 170 done                  \n",
      "Sample 180 done                  \n",
      "Sample 190 done                  \n",
      "Sample 200 done                  \n",
      "Sample 210 done                  \n",
      "Sample 220 done                  \n",
      "Sample 230 done                  \n",
      "Sample 240 done                  \n",
      "Sample 250 done                  \n",
      "Sample 260 done                  \n",
      "Sample 270 done                  \n",
      "Sample 280 done                  \n",
      "Sample 290 done                  \n",
      "Sample 300 done                  \n",
      "Sample 310 done                  \n",
      "Sample 320 done                  \n",
      "Sample 330 done                  \n",
      "Sample 340 done                  \n",
      "Sample 350 done                  \n",
      "Sample 360 done                  \n",
      "Sample 370 done                  \n",
      "Sample 380 done                  \n",
      "Sample 390 done                  \n",
      "Sample 400 done                  \n",
      "Sample 410 done                  \n",
      "Sample 420 done                  \n",
      "Sample 430 done                  \n",
      "Sample 440 done                  \n",
      "Sample 450 done                  \n",
      "Sample 460 done                  \n",
      "Sample 470 done                  \n",
      "Sample 480 done                  \n",
      "Sample 490 done                  \n",
      "Sample 500 done                  \n",
      "Sample 510 done                  \n",
      "Sample 520 done                  \n",
      "Sample 530 done                  \n",
      "Sample 540 done                  \n",
      "Sample 550 done                  \n",
      "Sample 560 done                  \n",
      "Sample 570 done                  \n",
      "Sample 580 done                  \n",
      "Sample 590 done                  \n",
      "Sample 600 done                  \n",
      "Sample 610 done                  \n",
      "Sample 620 done                  \n",
      "Sample 630 done                  \n",
      "Sample 640 done                  \n",
      "Sample 650 done                  \n",
      "Sample 660 done                  \n",
      "Sample 670 done                  \n",
      "Sample 680 done                  \n",
      "Sample 690 done                  \n",
      "Sample 700 done                  \n",
      "Sample 710 done                  \n",
      "Sample 720 done                  \n",
      "Sample 730 done                  \n",
      "Sample 740 done                  \n",
      "Sample 750 done                  \n",
      "Sample 760 done                  \n",
      "Done for cohort phase1 on chrom 2L\n"
     ]
    }
   ],
   "source": [
    "for chrom in ['X', '3R', '3L', '2R', '2L']:\n",
    "    for pop in sample_idx_dict.keys():\n",
    "        run_analysis(chrom, pop, sample_idx=sample_idx_dict[pop])\n",
    "        print(f'Done for cohort {pop} on chrom {chrom}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93c94cba-0ca5-4663-a446-c41fb1262c05",
   "metadata": {},
   "source": [
    "### Sanity check total number of non-variant targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5f5b75a2-1a92-40ab-9532-70e9a8358618",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_is_variant_all(chrom, sample_idx):\n",
    "    \"\"\"Locate variant sites for a given sample.\"\"\"\n",
    "    seq = get_reference_sequence(chrom)\n",
    "    out = np.zeros_like(seq, dtype=bool)\n",
    "    pos = ag3.snp_calls(chrom, sample_sets='3.0').variant_position.values\n",
    "    g = ag3.snp_calls(chrom, sample_sets = ['AG1000G-AO', 'AG1000G-GW', \n",
    "                                            'AG1000G-BF-A', 'AG1000G-GN-A', \n",
    "               'AG1000G-CM-A', 'AG1000G-GA-A', 'AG1000G-UG', 'AG1000G-KE'], \n",
    "                      sample_indices = sample_idx).call_genotype\n",
    "    loc = np.any(g > 0, axis=(1,2))\n",
    "    loc = loc.compute().values\n",
    "    idx = pos[loc] - 1\n",
    "    out[idx] = True\n",
    "\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "633c7b1f-e550-47d3-8b1c-ed9d3cf052df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def identify_targets_population_total(chrom, sample_idx):\n",
    "\n",
    "    #Find targets in cds (not using site filter)\n",
    "    base_loc = get_target_in_cds(chrom) \n",
    "    loc = base_loc & get_is_cas9_target(chrom, revcomp=True)\n",
    "    \n",
    "    #index genes\n",
    "    gene_idx = get_gene_idx(chrom)\n",
    "    \n",
    "    # setup outputs\n",
    "    out_sites = []\n",
    "    out_targets = []\n",
    "    out_genes = []\n",
    "    \n",
    "    # setup variation\n",
    "    is_variant = np.zeros(gene_idx.shape[0], dtype=bool)\n",
    "    \n",
    "    #initial numbers\n",
    "    out_sites.append(0)\n",
    "    \n",
    "    n_targets = np.count_nonzero(loc)\n",
    "    out_targets.append(n_targets)\n",
    "    \n",
    "    n_genes = len(np.unique(gene_idx[loc]))\n",
    "    out_genes.append(n_genes)\n",
    "        \n",
    "    #accumulate variation\n",
    "    is_variant |= get_is_variant_all(chrom, sample_idx)\n",
    "        \n",
    "    # locate sites\n",
    "    n_sites = np.count_nonzero(is_variant)\n",
    "    out_sites.append(n_sites)\n",
    "        \n",
    "    # locate targets and genes\n",
    "    loc = base_loc & get_is_cas9_target(chrom, revcomp=True, is_variant=is_variant)\n",
    "    n_targets = np.count_nonzero(loc)\n",
    "    out_targets.append(n_targets)\n",
    "\n",
    "    n_genes = len(np.unique(gene_idx[loc]))\n",
    "    out_genes.append(n_genes)\n",
    "\n",
    "\n",
    "    return np.array([out_sites, out_targets, out_genes]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6f8e00f4-ebb7-4532-b6de-e50f2a58c824",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 8,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 15,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29,\n",
       " 30,\n",
       " 31,\n",
       " 33,\n",
       " 34,\n",
       " 35,\n",
       " 37,\n",
       " 38,\n",
       " 39,\n",
       " 40,\n",
       " 41,\n",
       " 42,\n",
       " 43,\n",
       " 44,\n",
       " 46,\n",
       " 47,\n",
       " 49,\n",
       " 51,\n",
       " 53,\n",
       " 54,\n",
       " 55,\n",
       " 56,\n",
       " 57,\n",
       " 58,\n",
       " 59,\n",
       " 69,\n",
       " 70,\n",
       " 71,\n",
       " 72,\n",
       " 74,\n",
       " 75,\n",
       " 77,\n",
       " 79,\n",
       " 80,\n",
       " 81,\n",
       " 83,\n",
       " 84,\n",
       " 85,\n",
       " 86,\n",
       " 87,\n",
       " 88,\n",
       " 89,\n",
       " 90,\n",
       " 92,\n",
       " 93,\n",
       " 94,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 104,\n",
       " 107,\n",
       " 109,\n",
       " 111,\n",
       " 112,\n",
       " 113,\n",
       " 114,\n",
       " 115,\n",
       " 116,\n",
       " 117,\n",
       " 118,\n",
       " 121,\n",
       " 122,\n",
       " 123,\n",
       " 124,\n",
       " 125,\n",
       " 126,\n",
       " 127,\n",
       " 128,\n",
       " 129,\n",
       " 130,\n",
       " 132,\n",
       " 133,\n",
       " 134,\n",
       " 135,\n",
       " 137,\n",
       " 138,\n",
       " 139,\n",
       " 140,\n",
       " 141,\n",
       " 142,\n",
       " 143,\n",
       " 144,\n",
       " 146,\n",
       " 148,\n",
       " 150,\n",
       " 151,\n",
       " 152,\n",
       " 153,\n",
       " 154,\n",
       " 155,\n",
       " 156,\n",
       " 157,\n",
       " 158,\n",
       " 159,\n",
       " 160,\n",
       " 162,\n",
       " 163,\n",
       " 164,\n",
       " 165,\n",
       " 166,\n",
       " 167,\n",
       " 168,\n",
       " 169,\n",
       " 170,\n",
       " 171,\n",
       " 172,\n",
       " 173,\n",
       " 177,\n",
       " 178,\n",
       " 179,\n",
       " 181,\n",
       " 182,\n",
       " 183,\n",
       " 184,\n",
       " 185,\n",
       " 186,\n",
       " 187,\n",
       " 188,\n",
       " 189,\n",
       " 190,\n",
       " 191,\n",
       " 192,\n",
       " 193,\n",
       " 197,\n",
       " 199,\n",
       " 200,\n",
       " 201,\n",
       " 202,\n",
       " 203,\n",
       " 205,\n",
       " 206,\n",
       " 207,\n",
       " 208,\n",
       " 210,\n",
       " 212,\n",
       " 213,\n",
       " 214,\n",
       " 215,\n",
       " 216,\n",
       " 217,\n",
       " 218,\n",
       " 219,\n",
       " 220,\n",
       " 221,\n",
       " 222,\n",
       " 223,\n",
       " 224,\n",
       " 227,\n",
       " 228,\n",
       " 229,\n",
       " 230,\n",
       " 231,\n",
       " 233,\n",
       " 234,\n",
       " 235,\n",
       " 237,\n",
       " 238,\n",
       " 239,\n",
       " 240,\n",
       " 241,\n",
       " 242,\n",
       " 243,\n",
       " 244,\n",
       " 245,\n",
       " 247,\n",
       " 248,\n",
       " 249,\n",
       " 250,\n",
       " 251,\n",
       " 253,\n",
       " 254,\n",
       " 255,\n",
       " 256,\n",
       " 257,\n",
       " 258,\n",
       " 259,\n",
       " 260,\n",
       " 261,\n",
       " 262,\n",
       " 263,\n",
       " 264,\n",
       " 265,\n",
       " 266,\n",
       " 267,\n",
       " 269,\n",
       " 271,\n",
       " 272,\n",
       " 273,\n",
       " 274,\n",
       " 275,\n",
       " 277,\n",
       " 278,\n",
       " 279,\n",
       " 280,\n",
       " 281,\n",
       " 282,\n",
       " 283,\n",
       " 284,\n",
       " 285,\n",
       " 286,\n",
       " 287,\n",
       " 288,\n",
       " 289,\n",
       " 290,\n",
       " 291,\n",
       " 292,\n",
       " 293,\n",
       " 294,\n",
       " 295,\n",
       " 296,\n",
       " 297,\n",
       " 298,\n",
       " 300,\n",
       " 301,\n",
       " 302,\n",
       " 303,\n",
       " 304,\n",
       " 305,\n",
       " 306,\n",
       " 307,\n",
       " 308,\n",
       " 309,\n",
       " 310,\n",
       " 311,\n",
       " 312,\n",
       " 313,\n",
       " 314,\n",
       " 317,\n",
       " 318,\n",
       " 319,\n",
       " 320,\n",
       " 321,\n",
       " 322,\n",
       " 323,\n",
       " 324,\n",
       " 325,\n",
       " 326,\n",
       " 327,\n",
       " 328,\n",
       " 329,\n",
       " 330,\n",
       " 331,\n",
       " 333,\n",
       " 334,\n",
       " 335,\n",
       " 336,\n",
       " 337,\n",
       " 338,\n",
       " 339,\n",
       " 340,\n",
       " 341,\n",
       " 342,\n",
       " 343,\n",
       " 344,\n",
       " 345,\n",
       " 346,\n",
       " 347,\n",
       " 348,\n",
       " 349,\n",
       " 350,\n",
       " 351,\n",
       " 352,\n",
       " 353,\n",
       " 354,\n",
       " 355,\n",
       " 356,\n",
       " 357,\n",
       " 358,\n",
       " 359,\n",
       " 360,\n",
       " 361,\n",
       " 362,\n",
       " 363,\n",
       " 364,\n",
       " 365,\n",
       " 366,\n",
       " 367,\n",
       " 368,\n",
       " 369,\n",
       " 372,\n",
       " 373,\n",
       " 374,\n",
       " 375,\n",
       " 376,\n",
       " 377,\n",
       " 378,\n",
       " 379,\n",
       " 380,\n",
       " 381,\n",
       " 382,\n",
       " 383,\n",
       " 384,\n",
       " 386,\n",
       " 387,\n",
       " 388,\n",
       " 389,\n",
       " 390,\n",
       " 391,\n",
       " 392,\n",
       " 393,\n",
       " 395,\n",
       " 398,\n",
       " 400,\n",
       " 401,\n",
       " 402,\n",
       " 403,\n",
       " 404,\n",
       " 405,\n",
       " 406,\n",
       " 407,\n",
       " 408,\n",
       " 409,\n",
       " 410,\n",
       " 411,\n",
       " 412,\n",
       " 413,\n",
       " 414,\n",
       " 415,\n",
       " 416,\n",
       " 417,\n",
       " 418,\n",
       " 419,\n",
       " 420,\n",
       " 421,\n",
       " 422,\n",
       " 423,\n",
       " 424,\n",
       " 425,\n",
       " 426,\n",
       " 427,\n",
       " 428,\n",
       " 429,\n",
       " 430,\n",
       " 431,\n",
       " 432,\n",
       " 433,\n",
       " 434,\n",
       " 435,\n",
       " 436,\n",
       " 437,\n",
       " 438,\n",
       " 439,\n",
       " 440,\n",
       " 441,\n",
       " 442,\n",
       " 443,\n",
       " 445,\n",
       " 446,\n",
       " 447,\n",
       " 448,\n",
       " 449,\n",
       " 450,\n",
       " 451,\n",
       " 452,\n",
       " 453,\n",
       " 454,\n",
       " 455,\n",
       " 456,\n",
       " 457,\n",
       " 458,\n",
       " 459,\n",
       " 460,\n",
       " 461,\n",
       " 462,\n",
       " 463,\n",
       " 464,\n",
       " 465,\n",
       " 466,\n",
       " 467,\n",
       " 468,\n",
       " 469,\n",
       " 470,\n",
       " 471,\n",
       " 472,\n",
       " 473,\n",
       " 474,\n",
       " 475,\n",
       " 476,\n",
       " 477,\n",
       " 478,\n",
       " 480,\n",
       " 481,\n",
       " 482,\n",
       " 483,\n",
       " 484,\n",
       " 485,\n",
       " 486,\n",
       " 487,\n",
       " 488,\n",
       " 489,\n",
       " 490,\n",
       " 491,\n",
       " 492,\n",
       " 493,\n",
       " 494,\n",
       " 495,\n",
       " 496,\n",
       " 497,\n",
       " 498,\n",
       " 499,\n",
       " 500,\n",
       " 501,\n",
       " 502,\n",
       " 503,\n",
       " 504,\n",
       " 505,\n",
       " 506,\n",
       " 507,\n",
       " 508,\n",
       " 509,\n",
       " 511,\n",
       " 512,\n",
       " 513,\n",
       " 514,\n",
       " 515,\n",
       " 517,\n",
       " 518,\n",
       " 519,\n",
       " 520,\n",
       " 523,\n",
       " 525,\n",
       " 526,\n",
       " 527,\n",
       " 528,\n",
       " 529,\n",
       " 530,\n",
       " 531,\n",
       " 533,\n",
       " 534,\n",
       " 535,\n",
       " 537,\n",
       " 538,\n",
       " 539,\n",
       " 540,\n",
       " 541,\n",
       " 542,\n",
       " 543,\n",
       " 544,\n",
       " 546,\n",
       " 547,\n",
       " 548,\n",
       " 550,\n",
       " 551,\n",
       " 552,\n",
       " 553,\n",
       " 555,\n",
       " 556,\n",
       " 557,\n",
       " 558,\n",
       " 560,\n",
       " 561,\n",
       " 562,\n",
       " 564,\n",
       " 565,\n",
       " 566,\n",
       " 567,\n",
       " 568,\n",
       " 570,\n",
       " 571,\n",
       " 572,\n",
       " 573,\n",
       " 574,\n",
       " 575,\n",
       " 576,\n",
       " 577,\n",
       " 578,\n",
       " 579,\n",
       " 580,\n",
       " 581,\n",
       " 582,\n",
       " 583,\n",
       " 584,\n",
       " 585,\n",
       " 586,\n",
       " 588,\n",
       " 589,\n",
       " 591,\n",
       " 593,\n",
       " 594,\n",
       " 595,\n",
       " 596,\n",
       " 597,\n",
       " 598,\n",
       " 599,\n",
       " 600,\n",
       " 603,\n",
       " 604,\n",
       " 605,\n",
       " 607,\n",
       " 609,\n",
       " 611,\n",
       " 612,\n",
       " 613,\n",
       " 614,\n",
       " 615,\n",
       " 617,\n",
       " 618,\n",
       " 621,\n",
       " 622,\n",
       " 623,\n",
       " 624,\n",
       " 625,\n",
       " 626,\n",
       " 627,\n",
       " 628,\n",
       " 629,\n",
       " 630,\n",
       " 631,\n",
       " 632,\n",
       " 635,\n",
       " 638,\n",
       " 640,\n",
       " 641,\n",
       " 642,\n",
       " 643,\n",
       " 644,\n",
       " 645,\n",
       " 646,\n",
       " 647,\n",
       " 648,\n",
       " 649,\n",
       " 650,\n",
       " 651,\n",
       " 652,\n",
       " 653,\n",
       " 654,\n",
       " 655,\n",
       " 656,\n",
       " 659,\n",
       " 660,\n",
       " 661,\n",
       " 662,\n",
       " 663,\n",
       " 664,\n",
       " 672,\n",
       " 673,\n",
       " 674,\n",
       " 676,\n",
       " 677,\n",
       " 678,\n",
       " 679,\n",
       " 680,\n",
       " 681,\n",
       " 682,\n",
       " 683,\n",
       " 685,\n",
       " 686,\n",
       " 687,\n",
       " 688,\n",
       " 689,\n",
       " 690,\n",
       " 691,\n",
       " 692,\n",
       " 694,\n",
       " 695,\n",
       " 696,\n",
       " 697,\n",
       " 698,\n",
       " 699,\n",
       " 700,\n",
       " 701,\n",
       " 702,\n",
       " 703,\n",
       " 704,\n",
       " 706,\n",
       " 707,\n",
       " 708,\n",
       " 709,\n",
       " 710,\n",
       " 711,\n",
       " 712,\n",
       " 713,\n",
       " 715,\n",
       " 716,\n",
       " 717,\n",
       " 718,\n",
       " 719,\n",
       " 720,\n",
       " 721,\n",
       " 722,\n",
       " 723,\n",
       " 724,\n",
       " 725,\n",
       " 726,\n",
       " 729,\n",
       " 730,\n",
       " 795,\n",
       " 796,\n",
       " 797,\n",
       " 798,\n",
       " 799,\n",
       " 801,\n",
       " 802,\n",
       " 803,\n",
       " 804,\n",
       " 805,\n",
       " 806,\n",
       " 807,\n",
       " 808,\n",
       " 810,\n",
       " 811,\n",
       " 812,\n",
       " 813,\n",
       " 814,\n",
       " 815,\n",
       " 816,\n",
       " 817,\n",
       " 818,\n",
       " 819,\n",
       " 820,\n",
       " 821,\n",
       " 822,\n",
       " 823,\n",
       " 824,\n",
       " 825,\n",
       " 826,\n",
       " 828,\n",
       " 829,\n",
       " 830,\n",
       " 832,\n",
       " 833,\n",
       " 837,\n",
       " 838,\n",
       " 839,\n",
       " 840,\n",
       " 841,\n",
       " 842,\n",
       " 843,\n",
       " 948,\n",
       " 949,\n",
       " 950,\n",
       " 951,\n",
       " 952,\n",
       " 953,\n",
       " 954,\n",
       " 955,\n",
       " 956,\n",
       " 957,\n",
       " 958,\n",
       " 959,\n",
       " 960,\n",
       " 961,\n",
       " 962,\n",
       " 964,\n",
       " 965,\n",
       " 966,\n",
       " 967,\n",
       " 968,\n",
       " 969,\n",
       " 970,\n",
       " 971,\n",
       " 972,\n",
       " 973,\n",
       " 974,\n",
       " 975,\n",
       " 976,\n",
       " 977,\n",
       " 978,\n",
       " 979,\n",
       " 980,\n",
       " 981,\n",
       " 982,\n",
       " 983,\n",
       " 984,\n",
       " 985,\n",
       " 986,\n",
       " 987,\n",
       " 988,\n",
       " 989,\n",
       " 990,\n",
       " 991,\n",
       " 993,\n",
       " 994,\n",
       " 995,\n",
       " 996,\n",
       " 997,\n",
       " 998,\n",
       " 1000,\n",
       " 1001,\n",
       " 1002,\n",
       " 1003,\n",
       " 1005,\n",
       " 1006,\n",
       " 1007,\n",
       " 1008,\n",
       " 1009,\n",
       " 1010,\n",
       " 1011,\n",
       " 1012,\n",
       " 1014,\n",
       " 1016,\n",
       " 1017,\n",
       " 1018,\n",
       " 1019,\n",
       " 1020,\n",
       " 1021,\n",
       " 1022,\n",
       " 1023,\n",
       " 1024,\n",
       " 1025,\n",
       " 1026,\n",
       " 1027,\n",
       " 1028,\n",
       " 1029,\n",
       " 1030,\n",
       " 1031,\n",
       " 1032,\n",
       " 1034,\n",
       " 1035,\n",
       " 1036,\n",
       " 1037,\n",
       " 1038,\n",
       " 1039,\n",
       " 1040,\n",
       " 1042,\n",
       " 1043,\n",
       " 1044,\n",
       " 1045,\n",
       " 1046,\n",
       " 1047,\n",
       " 1048,\n",
       " 1049,\n",
       " 1050,\n",
       " 1051,\n",
       " 1052,\n",
       " 1053,\n",
       " 1055,\n",
       " 1056,\n",
       " 1057,\n",
       " 1058,\n",
       " 1059]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_idx = meta_p1.index.to_list()\n",
    "sample_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e5385daf-653e-4172-a45e-7afa9018462b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results_gam_no_site_filter/phase1//target_info_X.npy already exists, skipping computation\n",
      "Computing for 762 samples on chrom 3R\n",
      "Done for chrom 3R, total 5474 targets\n",
      "Computing for 762 samples on chrom 3L\n",
      "Done for chrom 3L, total 7462 targets\n",
      "Computing for 762 samples on chrom 2R\n",
      "Done for chrom 2R, total 7183 targets\n",
      "Computing for 762 samples on chrom 2L\n",
      "Done for chrom 2L, total 6528 targets\n"
     ]
    }
   ],
   "source": [
    "outdir = f'results_gam_no_site_filter/phase1/'\n",
    "for chrom in ['X', '3R', '3L', '2R', '2L']:\n",
    "    if os.path.exists(f'{outdir}/target_info_{chrom}.npy'): \n",
    "        print(f'{outdir}/target_info_{chrom}.npy already exists, skipping computation')\n",
    "    else:\n",
    "        if not os.path.isdir(outdir):\n",
    "            !mkdir {outdir}\n",
    "        \n",
    "        print(f'Computing for {len(sample_idx)} samples on chrom {chrom}')\n",
    "        target_info = identify_targets_population_total(chrom, sample_idx)\n",
    "        np.save(f'{outdir}/target_info_{chrom}.npy', target_info)\n",
    "        print(f'Done for chrom {chrom}, total {target_info[-1,1]} targets')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fdd5d7a1-133a-45ab-9f31-8821414caf96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[       0,   283999,     1059],\n",
       "       [11200479,     1774,      179]])"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.load('results_gam_no_site_filter/phase1/target_info_X.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "01583d70-755e-4167-8148-0e506b87d892",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28421"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1774+5474+7462+7183+6528"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d1e852ae-335a-4ead-8463-f6dc4df583fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22572d0d-3e3c-49b6-9bdc-c28f77d49c28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2584261b-fe90-4594-b6c1-b1a52950e122",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b86e51-5bd3-42c0-8051-2c9cdda65750",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a21a531d-02d5-46f4-a9df-71ec4ce5dc18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2acf2330-e6ba-4c05-92ae-78c089a846a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87145490-a51c-4588-8073-8ffbed828178",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "global-global-mgenv-6.0.6",
   "language": "python",
   "name": "conda-env-global-global-mgenv-6.0.6-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
